<!DOCTYPE HTML PUBLIC "-//W3C//DTD HTML 4.01 Transitional//EN"
                "http://www.w3.org/TR/REC-html40/loose.dtd">
<html>
<head>
  <title>Description of getMetaCycModelForOrganism</title>
  <meta name="keywords" content="getMetaCycModelForOrganism">
  <meta name="description" content="getMetaCycModelForOrganism">
  <meta http-equiv="Content-Type" content="text/html; charset=iso-8859-1">
  <meta name="generator" content="m2html v1.5 &copy; 2003-2005 Guillaume Flandin">
  <meta name="robots" content="index, follow">
  <link type="text/css" rel="stylesheet" href="../../m2html.css">
</head>
<body>
<a name="_top"></a>
<div><a href="../../index.html">Home</a> &gt;  <a href="../index.html">external</a> &gt; <a href="index.html">metacyc</a> &gt; getMetaCycModelForOrganism.m</div>

<!--<table width="100%"><tr><td align="left"><a href="../../index.html"><img alt="<" border="0" src="../../left.png">&nbsp;Master index</a></td>
<td align="right"><a href="index.html">Index for external\metacyc&nbsp;<img alt=">" border="0" src="../../right.png"></a></td></tr></table>-->

<h1>getMetaCycModelForOrganism
</h1>

<h2><a name="_name"></a>PURPOSE <a href="#_top"><img alt="^" border="0" src="../../up.png"></a></h2>
<div class="box"><strong>getMetaCycModelForOrganism</strong></div>

<h2><a name="_synopsis"></a>SYNOPSIS <a href="#_top"><img alt="^" border="0" src="../../up.png"></a></h2>
<div class="box"><strong>function model=getMetaCycModelForOrganism(organismID,fastaFile,keepTransportRxns,keepUnbalanced,keepUndetermined,minScore,minPositives,useDiamond) </strong></div>

<h2><a name="_description"></a>DESCRIPTION <a href="#_top"><img alt="^" border="0" src="../../up.png"></a></h2>
<div class="fragment"><pre class="comment"> getMetaCycModelForOrganism
   Reconstructs a genome-scale metabolic model based on protein homology to the
   MetaCyc pathway database

   Input:
   organismID          the query organism's abbreviation, which is defined
                       by user
   fastaFile           a FASTA file that contains the protein sequences of
                       the organism for which to reconstruct a model
   keepTransportRxns   include transportation reactions, which often have identical
                       reactants and products that turn to be all-zero columns in
                       the S matrix (opt, default false)
   keepUnbalanced      include reactions cannot be unbalanced reactions, usually
                       because they are polymeric reactions or because of a
                       specific difficulty in balancing class structures
                       (opt, default false)
   keepUndetermined    include reactions that have substrates lack chemical
                       structures or with non-numerical coefficients (e.g. n+1)
                       (opt, default false)
   minScore            minimum Bit scores of BLASTp search (opt, default 100)
   minPositives        minimum Positives values of BLASTp search (opt, default 45 %)
   useDiamond          use DIAMOND alignment tools to perform homology search
                       if true, otherwise the BLASTP is used (opt, default true)

   Output:
   model               a model structure for the query organism

   Usage: model=getMetaCycModelForOrganism(organismID,fastaFile,...
    keepTransportRxns,keepUnbalanced,keepUndetermined,minScore,minPositives,useDiamond)</pre></div>

<!-- crossreference -->
<h2><a name="_cross"></a>CROSS-REFERENCE INFORMATION <a href="#_top"><img alt="^" border="0" src="../../up.png"></a></h2>
This function calls:
<ul style="list-style-image:url(../../matlabicon.gif)">
<li><a href="getMetsFromMetaCyc.html" class="code" title="function metaCycMets=getMetsFromMetaCyc(metacycPath)">getMetsFromMetaCyc</a>	getMetsFromMetaCyc</li><li><a href="getModelFromMetaCyc.html" class="code" title="function metaCycModel=getModelFromMetaCyc(metacycPath,keepTransportRxns,keepUnbalanced,keepUndetermined)">getModelFromMetaCyc</a>	getModelFromMetaCyc</li></ul>
This function is called by:
<ul style="list-style-image:url(../../matlabicon.gif)">
</ul>
<!-- crossreference -->



<h2><a name="_source"></a>SOURCE CODE <a href="#_top"><img alt="^" border="0" src="../../up.png"></a></h2>
<div class="fragment"><pre>0001 <a name="_sub0" href="#_subfunctions" class="code">function model=getMetaCycModelForOrganism(organismID,fastaFile,</a><span class="keyword">...</span>
0002     keepTransportRxns,keepUnbalanced,keepUndetermined,minScore,minPositives,useDiamond)
0003 <span class="comment">% getMetaCycModelForOrganism</span>
0004 <span class="comment">%   Reconstructs a genome-scale metabolic model based on protein homology to the</span>
0005 <span class="comment">%   MetaCyc pathway database</span>
0006 <span class="comment">%</span>
0007 <span class="comment">%   Input:</span>
0008 <span class="comment">%   organismID          the query organism's abbreviation, which is defined</span>
0009 <span class="comment">%                       by user</span>
0010 <span class="comment">%   fastaFile           a FASTA file that contains the protein sequences of</span>
0011 <span class="comment">%                       the organism for which to reconstruct a model</span>
0012 <span class="comment">%   keepTransportRxns   include transportation reactions, which often have identical</span>
0013 <span class="comment">%                       reactants and products that turn to be all-zero columns in</span>
0014 <span class="comment">%                       the S matrix (opt, default false)</span>
0015 <span class="comment">%   keepUnbalanced      include reactions cannot be unbalanced reactions, usually</span>
0016 <span class="comment">%                       because they are polymeric reactions or because of a</span>
0017 <span class="comment">%                       specific difficulty in balancing class structures</span>
0018 <span class="comment">%                       (opt, default false)</span>
0019 <span class="comment">%   keepUndetermined    include reactions that have substrates lack chemical</span>
0020 <span class="comment">%                       structures or with non-numerical coefficients (e.g. n+1)</span>
0021 <span class="comment">%                       (opt, default false)</span>
0022 <span class="comment">%   minScore            minimum Bit scores of BLASTp search (opt, default 100)</span>
0023 <span class="comment">%   minPositives        minimum Positives values of BLASTp search (opt, default 45 %)</span>
0024 <span class="comment">%   useDiamond          use DIAMOND alignment tools to perform homology search</span>
0025 <span class="comment">%                       if true, otherwise the BLASTP is used (opt, default true)</span>
0026 <span class="comment">%</span>
0027 <span class="comment">%   Output:</span>
0028 <span class="comment">%   model               a model structure for the query organism</span>
0029 <span class="comment">%</span>
0030 <span class="comment">%   Usage: model=getMetaCycModelForOrganism(organismID,fastaFile,...</span>
0031 <span class="comment">%    keepTransportRxns,keepUnbalanced,keepUndetermined,minScore,minPositives,useDiamond)</span>
0032 
0033 <span class="keyword">if</span> nargin&lt;2
0034     EM=<span class="string">'No query protein fasta file is specified'</span>;
0035     dispEM(EM);
0036 <span class="keyword">end</span>
0037 <span class="keyword">if</span> nargin&lt;3
0038     keepTransportRxns=false;
0039 <span class="keyword">end</span>
0040 <span class="keyword">if</span> nargin&lt;4
0041     keepUnbalanced=false;
0042 <span class="keyword">end</span>
0043 <span class="keyword">if</span> nargin&lt;5
0044     keepUndetermined=false;
0045 <span class="keyword">end</span>
0046 <span class="keyword">if</span> nargin&lt;6
0047     minScore=100;
0048 <span class="keyword">end</span>
0049 <span class="keyword">if</span> nargin&lt;7
0050     minPositives=45;
0051 <span class="keyword">end</span>
0052 <span class="keyword">if</span> nargin&lt;8
0053     useDiamond=true;
0054 <span class="keyword">end</span>
0055 
0056 <span class="keyword">if</span> isempty(fastaFile)
0057     error(<span class="string">'*** The query FASTA filename cannot be empty! ***'</span>);
0058 <span class="keyword">else</span>
0059     fprintf(<span class="string">'\nChecking existence of query FASTA file... '</span>);
0060     <span class="comment">%Check if query fasta exists</span>
0061     fastaFile=checkFileExistence(fastaFile,2); <span class="comment">%Copy file to temp dir</span>
0062     fprintf(<span class="string">'done\n'</span>);
0063 <span class="keyword">end</span>
0064 
0065 <span class="comment">%First generate the full MetaCyc model</span>
0066 metaCycModel=<a href="getModelFromMetaCyc.html" class="code" title="function metaCycModel=getModelFromMetaCyc(metacycPath,keepTransportRxns,keepUnbalanced,keepUndetermined)">getModelFromMetaCyc</a>([],keepTransportRxns,keepUnbalanced,keepUndetermined);
0067 fprintf(<span class="string">'The full MetaCyc model loaded\n'</span>);
0068 
0069 <span class="comment">%Create the draft model from MetaCyc super model model=metaCycModel;</span>
0070 model.id=organismID;
0071 model.name=<span class="string">'Generated by homology with MetaCyc database'</span>;
0072 model.rxns=metaCycModel.rxns;
0073 model.rxnNames=metaCycModel.rxnNames;
0074 model.eccodes=metaCycModel.eccodes;
0075 model.subSystems=metaCycModel.subSystems;
0076 model.rxnMiriams=metaCycModel.rxnMiriams;
0077 model.rxnReferences=metaCycModel.rxnReferences;
0078 model.lb=metaCycModel.lb;
0079 model.ub=metaCycModel.ub;
0080 model.rev=metaCycModel.rev;
0081 model.c=metaCycModel.c;
0082 model.equations=metaCycModel.equations;
0083 
0084 <span class="comment">%Get the root directory for RAVEN Toolbox.</span>
0085 ravenPath=findRAVENroot();
0086 
0087 <span class="comment">%Generate blast strcture by either DIAMOND or BLASTP</span>
0088 <span class="keyword">if</span> useDiamond
0089     blastStruc=getDiamond(organismID,fastaFile,{<span class="string">'MetaCyc'</span>},fullfile(ravenPath,<span class="string">'external'</span>,<span class="string">'metacyc'</span>,<span class="string">'protseq.fsa'</span>));
0090 <span class="keyword">else</span>
0091     blastStruc=getBlast(organismID,fastaFile,{<span class="string">'MetaCyc'</span>},fullfile(ravenPath,<span class="string">'external'</span>,<span class="string">'metacyc'</span>,<span class="string">'protseq.fsa'</span>));
0092 <span class="keyword">end</span>
0093 
0094 <span class="comment">%Only look the query</span>
0095 blastStructure=blastStruc(2);
0096 
0097 <span class="comment">%Remove all blast hits that are below the cutoffs</span>
0098 indexes=blastStructure.bitscore&gt;=minScore &amp; blastStructure.ppos&gt;=minPositives;
0099 blastStructure.toGenes(~indexes)=[];
0100 blastStructure.fromGenes(~indexes)=[];
0101 blastStructure.evalue(~indexes)=[];
0102 blastStructure.aligLen(~indexes)=[];
0103 blastStructure.identity(~indexes)=[];
0104 blastStructure.bitscore(~indexes)=[];
0105 blastStructure.ppos(~indexes)=[];
0106 fprintf(<span class="string">'Completed searching against MetaCyc protein sequences.\n'</span>);
0107 
0108 <span class="comment">% Get the qualified genes of query organism from blast structure</span>
0109 model.genes=cell(10000,1);
0110 model.proteins=cell(10000,1);
0111 model.bitscore=zeros(10000,1);
0112 model.ppos=zeros(10000,1);
0113 num=1;
0114 
0115 <span class="comment">%Go through the strucutre and find out the hit with the best bit score</span>
0116 <span class="keyword">for</span> i=1:numel(blastStructure.fromGenes)
0117     <span class="keyword">if</span> num==1
0118         model.genes(num)=blastStructure.fromGenes(i);
0119         model.proteins(num)=strrep(blastStructure.toGenes(i), <span class="string">'gnl|META|'</span>, <span class="string">''</span>);
0120         model.bitscore(num)=blastStructure.bitscore(i);
0121         model.ppos(num)=blastStructure.ppos(i);
0122         num=num+1;
0123         lastGene=blastStructure.fromGenes(i);
0124     <span class="keyword">else</span>
0125         <span class="keyword">if</span> ~isequal(lastGene,blastStructure.fromGenes(i))
0126             model.genes(num)=blastStructure.fromGenes(i);
0127             model.proteins(num)=strrep(blastStructure.toGenes(i), <span class="string">'gnl|META|'</span>, <span class="string">''</span>);
0128             model.bitscore(num)=blastStructure.bitscore(i);
0129             model.ppos(num)=blastStructure.ppos(i);
0130             num=num+1;
0131             lastGene=blastStructure.fromGenes(i);
0132         <span class="keyword">else</span>
0133             <span class="keyword">if</span> model.bitscore(num)&lt;blastStructure.bitscore(i)
0134                 model.bitscore(num)=blastStructure.bitscore(i);
0135                 model.proteins(num)=strrep(blastStructure.toGenes(i), <span class="string">'gnl|META|'</span>, <span class="string">''</span>);
0136                 model.ppos(num)=blastStructure.ppos(i);
0137             <span class="keyword">end</span>
0138         <span class="keyword">end</span>
0139     <span class="keyword">end</span>
0140 <span class="keyword">end</span>
0141 model.genes=model.genes(1:num-1);
0142 model.proteins=model.proteins(1:num-1);
0143 model.bitscore=model.bitscore(1:num-1);
0144 model.ppos=model.ppos(1:num-1);
0145 
0146 <span class="comment">% Get the indexes of matched genes in the metaCycModel</span>
0147 <span class="comment">% because some enzymes in the FASTA file cannot be found in the dump file</span>
0148 [hits, indexes]=ismember(model.proteins,metaCycModel.genes);
0149 found = find(hits);
0150 model.genes=model.genes(found);
0151 
0152 <span class="comment">% Restructure the rxnGeneMat matrix</span>
0153 model.rxnGeneMat=sparse(metaCycModel.rxnGeneMat(:,indexes(found)));
0154 
0155 <span class="comment">%Remove all reactions without genes</span>
0156 hasGenes=any(model.rxnGeneMat,2);
0157 model=removeReactions(model,~hasGenes,true);
0158 
0159 <span class="comment">%Generate grRules, only consider the or relationship here Matched enzymes</span>
0160 <span class="comment">%are stored in field rxnScores,</span>
0161 rxnNum=numel(model.rxns);
0162 model.rxnConfidenceScores=NaN(rxnNum,1);
0163 model.rxnConfidenceScores(:)=2;
0164 model.grRules=cell(rxnNum,1);
0165 <span class="comment">%model.rxnScores=cell(rxnNum,1);</span>
0166 <span class="keyword">for</span> j=1:rxnNum
0167     model.grRules{j}=<span class="string">''</span>;
0168     I=find(model.rxnGeneMat(j,:));
0169     <span class="keyword">for</span> k=1:numel(I)
0170         <span class="keyword">if</span> isempty(model.grRules{j})
0171             model.grRules(j)=model.genes(I(k));
0172             <span class="comment">%model.rxnScores(j)=model.proteins(I(k));</span>
0173         <span class="keyword">else</span>
0174             model.grRules(j)=strcat(model.grRules(j),{<span class="string">' or '</span>},model.genes(I(k)));
0175             <span class="comment">%model.rxnScores(j)=strcat(model.rxnScores(j),{' or</span>
0176             <span class="comment">%'},model.proteins(I(k)));</span>
0177         <span class="keyword">end</span>
0178     <span class="keyword">end</span>
0179 <span class="keyword">end</span>
0180 <span class="comment">%update genes field</span>
0181 model.genes=model.genes(any(model.rxnGeneMat));
0182 
0183 <span class="comment">%Construct the S matrix and list of metabolites</span>
0184 [S, mets, badRxns]=constructS(model.equations);
0185 model.S=S;
0186 model.mets=mets;
0187 
0188 <span class="comment">%model=removeReactions(model,badRxns,true,true);</span>
0189 
0190 <span class="comment">%Then match up metabolites</span>
0191 metaCycMets=<a href="getMetsFromMetaCyc.html" class="code" title="function metaCycMets=getMetsFromMetaCyc(metacycPath)">getMetsFromMetaCyc</a>([]);
0192 
0193 <span class="comment">%Add information about all metabolites to the model</span>
0194 [a, b]=ismember(model.mets,metaCycMets.mets);
0195 a=find(a);
0196 b=b(a);
0197 
0198 <span class="keyword">if</span> ~isfield(model,<span class="string">'metNames'</span>)
0199     model.metNames=cell(numel(model.mets),1);
0200     model.metNames(:)={<span class="string">''</span>};
0201 <span class="keyword">end</span>
0202 model.metNames(a)=metaCycMets.metNames(b);
0203 
0204 <span class="keyword">if</span> ~isfield(model,<span class="string">'metFormulas'</span>)
0205     model.metFormulas=cell(numel(model.mets),1);
0206     model.metFormulas(:)={<span class="string">''</span>};
0207 <span class="keyword">end</span>
0208 model.metFormulas(a)=metaCycMets.metFormulas(b);
0209 
0210 <span class="keyword">if</span> ~isfield(model,<span class="string">'metCharges'</span>)
0211     model.metCharges=zeros(numel(model.mets),1);
0212 <span class="keyword">end</span>
0213 model.metCharges(a)=metaCycMets.metCharges(b);
0214 
0215 <span class="keyword">if</span> ~isfield(model,<span class="string">'b'</span>)
0216     model.b=zeros(numel(model.mets),1);
0217 <span class="keyword">end</span>
0218 <span class="comment">%model.b(a)=metaCycMets.b(b);</span>
0219 
0220 <span class="keyword">if</span> ~isfield(model,<span class="string">'inchis'</span>)
0221     model.inchis=cell(numel(model.mets),1);
0222     model.inchis(:)={<span class="string">''</span>};
0223 <span class="keyword">end</span>
0224 model.inchis(a)=metaCycMets.inchis(b);
0225 
0226 <span class="keyword">if</span> ~isfield(model,<span class="string">'metMiriams'</span>)
0227     model.metMiriams=cell(numel(model.mets),1);
0228 <span class="keyword">end</span>
0229 model.metMiriams(a)=metaCycMets.metMiriams(b);
0230 
0231 <span class="comment">%Put all metabolites in one compartment called 's' (for system). This is</span>
0232 <span class="comment">%done just to be more compatible with the rest of the code</span>
0233 model.comps={<span class="string">'s'</span>};
0234 model.compNames={<span class="string">'System'</span>};
0235 model.metComps=ones(numel(model.mets),1);
0236 
0237 <span class="comment">%It could also be that the metabolite names are empty for some reason In</span>
0238 <span class="comment">%that case, use the ID instead</span>
0239 I=cellfun(@isempty,model.metNames);
0240 model.metNames(I)=model.mets(I);
0241 
0242 <span class="comment">%Remove additional fields</span>
0243 model=rmfield(model,{<span class="string">'proteins'</span>,<span class="string">'bitscore'</span>,<span class="string">'ppos'</span>});
0244 
0245 <span class="comment">%In the end fix grRules and rxnGeneMat</span>
0246 [grRules,rxnGeneMat] = standardizeGrRules(model,false); <span class="comment">%Get detailed output</span>
0247 model.grRules = grRules;
0248 model.rxnGeneMat = rxnGeneMat;
0249 <span class="comment">%Remove the temp fasta file</span>
0250 delete(fastaFile)
0251 <span class="keyword">end</span></pre></div>
<hr><address>Generated by <strong><a href="http://www.artefact.tk/software/matlab/m2html/" title="Matlab Documentation in HTML">m2html</a></strong> &copy; 2005</address>
</body>
</html>