<!DOCTYPE HTML PUBLIC "-//W3C//DTD HTML 4.01 Transitional//EN"
                "http://www.w3.org/TR/REC-html40/loose.dtd">
<html>
<head>
  <title>Description of getWoLFScores</title>
  <meta name="keywords" content="getWoLFScores">
  <meta name="description" content="getWoLFScores">
  <meta http-equiv="Content-Type" content="text/html; charset=iso-8859-1">
  <meta name="generator" content="m2html v1.5 &copy; 2003-2005 Guillaume Flandin">
  <meta name="robots" content="index, follow">
  <link type="text/css" rel="stylesheet" href="../m2html.css">
</head>
<body>
<a name="_top"></a>
<div><a href="../index.html">Home</a> &gt;  <a href="index.html">external</a> &gt; getWoLFScores.m</div>

<!--<table width="100%"><tr><td align="left"><a href="../index.html"><img alt="<" border="0" src="../left.png">&nbsp;Master index</a></td>
<td align="right"><a href="index.html">Index for external&nbsp;<img alt=">" border="0" src="../right.png"></a></td></tr></table>-->

<h1>getWoLFScores
</h1>

<h2><a name="_name"></a>PURPOSE <a href="#_top"><img alt="^" border="0" src="../up.png"></a></h2>
<div class="box"><strong>getWoLFScores</strong></div>

<h2><a name="_synopsis"></a>SYNOPSIS <a href="#_top"><img alt="^" border="0" src="../up.png"></a></h2>
<div class="box"><strong>function GSS = getWoLFScores(inputFile, kingdom) </strong></div>

<h2><a name="_description"></a>DESCRIPTION <a href="#_top"><img alt="^" border="0" src="../up.png"></a></h2>
<div class="fragment"><pre class="comment"> getWoLFScores
   Call WoLF PSort to predict the sub-cellular localization of proteins.
   The output can be used as input to predictLocalization. This function
   is currently only available for Linux and requires Perl to be
   installed. If one wants to use another predictor, see parseScores. The
   function normalizes the scores so that the best score for each gene is
   1.0.

   Input:
   inputFile    a FASTA file with protein sequences
   kingdom     the kingdom of the organism, 'animal', 'fungi' or 'plant'

   Output:
   GSS         a gene scoring structure to be used in predictLocalization

   Usage: GSS = getWoLFScores(inputFile, kingdom)</pre></div>

<!-- crossreference -->
<h2><a name="_cross"></a>CROSS-REFERENCE INFORMATION <a href="#_top"><img alt="^" border="0" src="../up.png"></a></h2>
This function calls:
<ul style="list-style-image:url(../matlabicon.gif)">
<li><a href="parseScores.html" class="code" title="function GSS = parseScores(inputFile, predictor)">parseScores</a>	parseScores</li></ul>
This function is called by:
<ul style="list-style-image:url(../matlabicon.gif)">
</ul>
<!-- crossreference -->



<h2><a name="_source"></a>SOURCE CODE <a href="#_top"><img alt="^" border="0" src="../up.png"></a></h2>
<div class="fragment"><pre>0001 <a name="_sub0" href="#_subfunctions" class="code">function GSS = getWoLFScores(inputFile, kingdom)</a>
0002 <span class="comment">% getWoLFScores</span>
0003 <span class="comment">%   Call WoLF PSort to predict the sub-cellular localization of proteins.</span>
0004 <span class="comment">%   The output can be used as input to predictLocalization. This function</span>
0005 <span class="comment">%   is currently only available for Linux and requires Perl to be</span>
0006 <span class="comment">%   installed. If one wants to use another predictor, see parseScores. The</span>
0007 <span class="comment">%   function normalizes the scores so that the best score for each gene is</span>
0008 <span class="comment">%   1.0.</span>
0009 <span class="comment">%</span>
0010 <span class="comment">%   Input:</span>
0011 <span class="comment">%   inputFile    a FASTA file with protein sequences</span>
0012 <span class="comment">%   kingdom     the kingdom of the organism, 'animal', 'fungi' or 'plant'</span>
0013 <span class="comment">%</span>
0014 <span class="comment">%   Output:</span>
0015 <span class="comment">%   GSS         a gene scoring structure to be used in predictLocalization</span>
0016 <span class="comment">%</span>
0017 <span class="comment">%   Usage: GSS = getWoLFScores(inputFile, kingdom)</span>
0018 
0019 <span class="keyword">if</span> ~(exist(inputFile,<span class="string">'file'</span>)==2)
0020     error(<span class="string">'FASTA file %s cannot be found'</span>,string(inputFile));
0021 <span class="keyword">end</span>
0022 
0023 <span class="keyword">if</span> ~strcmp(kingdom,<span class="string">'animal'</span>) &amp;&amp; ~strcmp(kingdom,<span class="string">'fungi'</span>) &amp;&amp; ~strcmp(kingdom,<span class="string">'plant'</span>)
0024     EM=<span class="string">'Allowed kingdoms are &quot;animal&quot;, &quot;fungi&quot;, and &quot;plant&quot;'</span>;
0025     dispEM(EM);
0026 <span class="keyword">end</span>
0027 
0028 <span class="keyword">if</span> ispc || ismac
0029     EM=<span class="string">'This function currently runs only on Linux. Use parseScores if you want to use another predictor'</span>;
0030     dispEM(EM);
0031 <span class="keyword">end</span>
0032 
0033 <span class="comment">%Get the directory for RAVEN Toolbox</span>
0034 ravenPath=findRAVENroot();
0035 
0036 <span class="comment">%Temporary output name</span>
0037 outFile=tempname;
0038 fid=fopen(outFile,<span class="string">'w'</span>);
0039 
0040 <span class="comment">%Do the prediction</span>
0041 [~, output]=unix([<span class="string">'perl &quot;'</span> ravenPath <span class="string">'/software/WoLFPSORT/bin/runWolfPsortSummary&quot; '</span> kingdom <span class="string">' &lt; '</span> inputFile]);
0042 
0043 <span class="comment">%Save output and call the general parser</span>
0044 fprintf(fid,output);
0045 fclose(fid);
0046 GSS=<a href="parseScores.html" class="code" title="function GSS = parseScores(inputFile, predictor)">parseScores</a>(outFile,<span class="string">'wolf'</span>);
0047 
0048 <span class="comment">%Clean up</span>
0049 delete(outFile);
0050 <span class="keyword">end</span></pre></div>
<hr><address>Generated by <strong><a href="http://www.artefact.tk/software/matlab/m2html/" title="Matlab Documentation in HTML">m2html</a></strong> &copy; 2005</address>
</body>
</html>