<!DOCTYPE HTML PUBLIC "-//W3C//DTD HTML 4.01 Transitional//EN"
                "http://www.w3.org/TR/REC-html40/loose.dtd">
<html>
<head>
  <title>Description of getModelFromHomology</title>
  <meta name="keywords" content="getModelFromHomology">
  <meta name="description" content="getModelFromHomology">
  <meta http-equiv="Content-Type" content="text/html; charset=iso-8859-1">
  <meta name="generator" content="m2html v1.5 &copy; 2003-2005 Guillaume Flandin">
  <meta name="robots" content="index, follow">
  <link type="text/css" rel="stylesheet" href="../m2html.css">
</head>
<body>
<a name="_top"></a>
<div><a href="../index.html">Home</a> &gt;  <a href="index.html">core</a> &gt; getModelFromHomology.m</div>

<!--<table width="100%"><tr><td align="left"><a href="../index.html"><img alt="<" border="0" src="../left.png">&nbsp;Master index</a></td>
<td align="right"><a href="index.html">Index for core&nbsp;<img alt=">" border="0" src="../right.png"></a></td></tr></table>-->

<h1>getModelFromHomology
</h1>

<h2><a name="_name"></a>PURPOSE <a href="#_top"><img alt="^" border="0" src="../up.png"></a></h2>
<div class="box"><strong>getModelFromHomology</strong></div>

<h2><a name="_synopsis"></a>SYNOPSIS <a href="#_top"><img alt="^" border="0" src="../up.png"></a></h2>
<div class="box"><strong>function [draftModel, hitGenes]=getModelFromHomology(models,blastStructure,getModelFor,preferredOrder,strictness,onlyGenesInModels,maxE,minLen,minIde,mapNewGenesToOld) </strong></div>

<h2><a name="_description"></a>DESCRIPTION <a href="#_top"><img alt="^" border="0" src="../up.png"></a></h2>
<div class="fragment"><pre class="comment"> getModelFromHomology
   Constructs a new model from a set of existing models and gene homology
   information.

   models            a cell array of model structures to build the model
                     from. These models must be sorted by importance in
                     decreasing order
   blastStructure    a blastStructure as produced by getBlast or
                     getBlastFromExcel
   getModelFor       a three-four letter abbreviation of the organism to
                     build a model for. Must have BLASTP hits in both
                     directions to the organisms in 'models'
   preferredOrder    the order in which reactions should be added from the
                     models. If not supplied, reactions will be included
                     from all models, otherwise one gene will only result
                     in reactions from one model (opt, default {})
   strictness        integer that specifies which reactions should be
                     included:
                     1: Map new genes to old for all pairs, which have
                     acceptable BLASTP results in both directions
                     2: Map new genes to old for all pairs, which have
                     acceptable BLASTP results in correspondent direction
                     (mapping can be done in the opposite direction, see
                     mapNewGenesToOld below)
                     3: Check all BLASTP results and retain only the best
                     results by E-value for all gene pairs in each
                     direction separately. Then map new genes to old for
                     all pairs, which have acceptable BLASTP results in
                     both directions (opt, default 1).
   onlyGenesInModels consider BLASTP results only for genes that exist in
                     the models. This tends to import a larger fraction
                     from the existing models but may give less reliable
                     results. Has effect only if strictness=3 (opt,
                     default false)
   maxE              only look at genes with E-values &lt;= this value (opt,
                     default 10^-30)
   minLen            only look at genes with alignment length &gt;= this
                     value (opt, default 200)
   minIde            only look at genes with identity &gt;= this value
                     (opt, default 40 (%))
   mapNewGenesToOld  determines how to match genes if not looking at only
                     1-1 orthologs. Either map the new genes to the old or
                     old genes to new. The default is to map the new genes
                     (opt, default true)

   draftModel        a model structure for the new organism
   hitGenes          collect the old and new genes

   The models in the 'models' structure should have named the metabolites
   in the same manner, have their reversible reactions in the same
   direction (run sortModel), and use the same compartment names. To avoid
   keeping unneccesary old genes, the models should not have
   'or'-relations in their grRules (use expandModel).

   The resulting draft model contains only reactions associated with
   orthologous genes. The old (original) genes involved in 'and'
   relations in grRules without any orthologs are still included in
   the draft model as OLD_MODELID_geneName.

   NOTE: &quot;to&quot; and &quot;from&quot; means relative to the new organism

   Usage: draftModel=getModelFromHomology(models,blastStructure,...
    getModelFor,preferredOrder,strictness,onlyGenesInModels,maxE,...
    minLen,minIde,mapNewGenesToOld)</pre></div>

<!-- crossreference -->
<h2><a name="_cross"></a>CROSS-REFERENCE INFORMATION <a href="#_top"><img alt="^" border="0" src="../up.png"></a></h2>
This function calls:
<ul style="list-style-image:url(../matlabicon.gif)">
<li><a href="convertCharArray.html" class="code" title="function inputConverted = convertCharArray(funcInput)">convertCharArray</a>	convertCharArray</li><li><a href="deleteUnusedGenes.html" class="code" title="function reducedModel=deleteUnusedGenes(model,verbose)">deleteUnusedGenes</a>	deleteUnusedGenes</li><li><a href="dispEM.html" class="code" title="function dispEM(string,throwErrors,toList,trimWarnings)">dispEM</a>	dispEM</li><li><a href="mergeModels.html" class="code" title="function model=mergeModels(models,metParam,supressWarnings)">mergeModels</a>	mergeModels</li><li><a href="removeGenes.html" class="code" title="function reducedModel = removeGenes(model,genesToRemove,removeUnusedMets,removeBlockedRxns,standardizeRules)">removeGenes</a>	removeGenes</li><li><a href="removeReactions.html" class="code" title="function reducedModel=removeReactions(model,rxnsToRemove,removeUnusedMets,removeUnusedGenes,removeUnusedComps)">removeReactions</a>	removeReactions</li><li><a href="standardizeGrRules.html" class="code" title="function [grRules,rxnGeneMat,indexes2check] = standardizeGrRules(model,embedded)">standardizeGrRules</a>	standardizeGrRules</li></ul>
This function is called by:
<ul style="list-style-image:url(../matlabicon.gif)">
</ul>
<!-- crossreference -->



<h2><a name="_source"></a>SOURCE CODE <a href="#_top"><img alt="^" border="0" src="../up.png"></a></h2>
<div class="fragment"><pre>0001 <a name="_sub0" href="#_subfunctions" class="code">function [draftModel, hitGenes]=getModelFromHomology(models,blastStructure,</a><span class="keyword">...</span>
0002     getModelFor,preferredOrder,strictness,onlyGenesInModels,maxE,<span class="keyword">...</span>
0003     minLen,minIde,mapNewGenesToOld)
0004 <span class="comment">% getModelFromHomology</span>
0005 <span class="comment">%   Constructs a new model from a set of existing models and gene homology</span>
0006 <span class="comment">%   information.</span>
0007 <span class="comment">%</span>
0008 <span class="comment">%   models            a cell array of model structures to build the model</span>
0009 <span class="comment">%                     from. These models must be sorted by importance in</span>
0010 <span class="comment">%                     decreasing order</span>
0011 <span class="comment">%   blastStructure    a blastStructure as produced by getBlast or</span>
0012 <span class="comment">%                     getBlastFromExcel</span>
0013 <span class="comment">%   getModelFor       a three-four letter abbreviation of the organism to</span>
0014 <span class="comment">%                     build a model for. Must have BLASTP hits in both</span>
0015 <span class="comment">%                     directions to the organisms in 'models'</span>
0016 <span class="comment">%   preferredOrder    the order in which reactions should be added from the</span>
0017 <span class="comment">%                     models. If not supplied, reactions will be included</span>
0018 <span class="comment">%                     from all models, otherwise one gene will only result</span>
0019 <span class="comment">%                     in reactions from one model (opt, default {})</span>
0020 <span class="comment">%   strictness        integer that specifies which reactions should be</span>
0021 <span class="comment">%                     included:</span>
0022 <span class="comment">%                     1: Map new genes to old for all pairs, which have</span>
0023 <span class="comment">%                     acceptable BLASTP results in both directions</span>
0024 <span class="comment">%                     2: Map new genes to old for all pairs, which have</span>
0025 <span class="comment">%                     acceptable BLASTP results in correspondent direction</span>
0026 <span class="comment">%                     (mapping can be done in the opposite direction, see</span>
0027 <span class="comment">%                     mapNewGenesToOld below)</span>
0028 <span class="comment">%                     3: Check all BLASTP results and retain only the best</span>
0029 <span class="comment">%                     results by E-value for all gene pairs in each</span>
0030 <span class="comment">%                     direction separately. Then map new genes to old for</span>
0031 <span class="comment">%                     all pairs, which have acceptable BLASTP results in</span>
0032 <span class="comment">%                     both directions (opt, default 1).</span>
0033 <span class="comment">%   onlyGenesInModels consider BLASTP results only for genes that exist in</span>
0034 <span class="comment">%                     the models. This tends to import a larger fraction</span>
0035 <span class="comment">%                     from the existing models but may give less reliable</span>
0036 <span class="comment">%                     results. Has effect only if strictness=3 (opt,</span>
0037 <span class="comment">%                     default false)</span>
0038 <span class="comment">%   maxE              only look at genes with E-values &lt;= this value (opt,</span>
0039 <span class="comment">%                     default 10^-30)</span>
0040 <span class="comment">%   minLen            only look at genes with alignment length &gt;= this</span>
0041 <span class="comment">%                     value (opt, default 200)</span>
0042 <span class="comment">%   minIde            only look at genes with identity &gt;= this value</span>
0043 <span class="comment">%                     (opt, default 40 (%))</span>
0044 <span class="comment">%   mapNewGenesToOld  determines how to match genes if not looking at only</span>
0045 <span class="comment">%                     1-1 orthologs. Either map the new genes to the old or</span>
0046 <span class="comment">%                     old genes to new. The default is to map the new genes</span>
0047 <span class="comment">%                     (opt, default true)</span>
0048 <span class="comment">%</span>
0049 <span class="comment">%   draftModel        a model structure for the new organism</span>
0050 <span class="comment">%   hitGenes          collect the old and new genes</span>
0051 <span class="comment">%</span>
0052 <span class="comment">%   The models in the 'models' structure should have named the metabolites</span>
0053 <span class="comment">%   in the same manner, have their reversible reactions in the same</span>
0054 <span class="comment">%   direction (run sortModel), and use the same compartment names. To avoid</span>
0055 <span class="comment">%   keeping unneccesary old genes, the models should not have</span>
0056 <span class="comment">%   'or'-relations in their grRules (use expandModel).</span>
0057 <span class="comment">%</span>
0058 <span class="comment">%   The resulting draft model contains only reactions associated with</span>
0059 <span class="comment">%   orthologous genes. The old (original) genes involved in 'and'</span>
0060 <span class="comment">%   relations in grRules without any orthologs are still included in</span>
0061 <span class="comment">%   the draft model as OLD_MODELID_geneName.</span>
0062 <span class="comment">%</span>
0063 <span class="comment">%   NOTE: &quot;to&quot; and &quot;from&quot; means relative to the new organism</span>
0064 <span class="comment">%</span>
0065 <span class="comment">%   Usage: draftModel=getModelFromHomology(models,blastStructure,...</span>
0066 <span class="comment">%    getModelFor,preferredOrder,strictness,onlyGenesInModels,maxE,...</span>
0067 <span class="comment">%    minLen,minIde,mapNewGenesToOld)</span>
0068 
0069 hitGenes.oldGenes = [];  <span class="comment">% collect the old genes from the template model (organism)</span>
0070 hitGenes.newGenes = [];  <span class="comment">% collect the new genes of the draft model (target organism)</span>
0071 
0072 getModelFor=char(getModelFor);
0073 
0074 <span class="keyword">if</span> nargin&lt;4
0075     preferredOrder=[];
0076 <span class="keyword">else</span>
0077     preferredOrder=<a href="convertCharArray.html" class="code" title="function inputConverted = convertCharArray(funcInput)">convertCharArray</a>(preferredOrder);
0078 <span class="keyword">end</span>
0079 <span class="keyword">if</span> nargin&lt;5
0080     strictness=1;
0081 <span class="keyword">end</span>
0082 <span class="keyword">if</span> nargin&lt;6
0083     onlyGenesInModels=false;
0084 <span class="keyword">end</span>
0085 <span class="keyword">if</span> nargin&lt;7
0086     maxE=10^-30;
0087 <span class="keyword">end</span>
0088 <span class="keyword">if</span> nargin&lt;8
0089     minLen=200;
0090 <span class="keyword">end</span>
0091 <span class="keyword">if</span> nargin&lt;9
0092     minIde=40;
0093 <span class="keyword">end</span>
0094 <span class="keyword">if</span> nargin&lt;10
0095     mapNewGenesToOld=true;
0096 <span class="keyword">end</span>
0097 
0098 <span class="keyword">if</span> isfield(models,<span class="string">'S'</span>)
0099     models={models};
0100 <span class="keyword">end</span>
0101     
0102 <span class="comment">%Check that all the information is in the blast structure</span>
0103 modelNames=cell(numel(models),1);
0104 <span class="keyword">for</span> i=1:numel(models)
0105     modelNames{i}=models{i}.id;
0106     <span class="comment">%Gene short names and geneMiriams are often different between species,</span>
0107     <span class="comment">%safer not to include them</span>
0108     <span class="keyword">if</span> isfield(models{i},<span class="string">'geneShortNames'</span>)
0109         models{i}=rmfield(models{i},<span class="string">'geneShortNames'</span>);
0110     <span class="keyword">end</span>
0111     <span class="keyword">if</span> isfield(models{i},<span class="string">'geneMiriams'</span>)
0112         models{i}=rmfield(models{i},<span class="string">'geneMiriams'</span>);
0113     <span class="keyword">end</span>
0114 <span class="keyword">end</span>
0115 
0116 <span class="comment">%Assume for now that all information is there and that it's correct. This</span>
0117 <span class="comment">%is important to fix since no further checks are being made!</span>
0118 
0119 <span class="comment">%Check whether provided fasta files use the same gene identifiers as</span>
0120 <span class="comment">%provided template models</span>
0121 <span class="keyword">for</span> i=1:numel(blastStructure)
0122     <span class="keyword">if</span> ~strcmp(blastStructure(i).fromId,getModelFor)
0123         j=strcmpi(blastStructure(i).fromId,modelNames);
0124         <span class="keyword">if</span> j==0
0125             error([<span class="string">'While the blastStructure contains sequences from '</span><span class="keyword">...</span>
0126                 <span class="string">'organismID &quot;%s&quot; (as\nprovided in getBlast), none of '</span><span class="keyword">...</span>
0127                 <span class="string">'template models have this id (as model.id)'</span>],<span class="keyword">...</span>
0128                 string(blastStructure(i).fromId));
0129         <span class="keyword">end</span>
0130         k=sum(ismember(blastStructure(i).fromGenes,models{j}.genes));
0131         <span class="keyword">if</span> k&lt;(numel(models{j}.genes)*0.05)
0132             error([<span class="string">'Less than 5%% of the genes in the template model '</span><span class="keyword">...</span>
0133                 <span class="string">'with model.id &quot;%s&quot;\ncan be found in the blastStructure. '</span><span class="keyword">...</span>
0134                 <span class="string">'Ensure that the protein FASTA\nused in getBlast and '</span><span class="keyword">...</span>
0135                 <span class="string">'the template model used in getModelFromHomology\nuse '</span><span class="keyword">...</span>
0136                 <span class="string">'the same style of gene identifiers'</span>],models{j}.id)
0137         <span class="keyword">end</span>
0138     <span class="keyword">end</span>
0139 <span class="keyword">end</span>
0140 
0141 <span class="comment">%Standardize grRules of template models</span>
0142 <span class="keyword">for</span> i=1:length(models)
0143     fprintf(<span class="string">'\nStandardizing grRules of template model with ID &quot;%s&quot; ...'</span>,models{i}.id);
0144     [models{i}.grRules,models{i}.rxnGeneMat]=<a href="standardizeGrRules.html" class="code" title="function [grRules,rxnGeneMat,indexes2check] = standardizeGrRules(model,embedded)">standardizeGrRules</a>(models{i},false);
0145 <span class="keyword">end</span>
0146 fprintf(<span class="string">' done\n'</span>);
0147 
0148 <span class="comment">%Remove all gene matches that are below the cutoffs</span>
0149 <span class="keyword">for</span> i=1:numel(blastStructure)
0150     indexes=blastStructure(i).evalue&lt;maxE &amp; blastStructure(i).aligLen&gt;=minLen &amp; blastStructure(i).identity&gt;=minIde; <span class="comment">%Do it in this direction to lose NaNs</span>
0151     blastStructure(i).fromGenes(~indexes)=[];
0152     blastStructure(i).toGenes(~indexes)=[];
0153     blastStructure(i).evalue(~indexes)=[];
0154     blastStructure(i).identity(~indexes)=[];
0155     blastStructure(i).aligLen(~indexes)=[];
0156     blastStructure(i).bitscore(~indexes)=[];
0157     blastStructure(i).ppos(~indexes)=[];
0158 <span class="keyword">end</span>
0159 
0160 <span class="comment">%Remove all reactions from the models that have no genes encoding for them.</span>
0161 <span class="comment">%Also remove all genes that encode for no reactions. There shouldn't be any</span>
0162 <span class="comment">%but there might be mistakes</span>
0163 <span class="keyword">for</span> i=1:numel(models)
0164     [hasGenes, ~]=find(models{i}.rxnGeneMat);
0165     hasNoGenes=1:numel(models{i}.rxns);
0166     hasNoGenes(hasGenes)=[];
0167     models{i}=<a href="removeReactions.html" class="code" title="function reducedModel=removeReactions(model,rxnsToRemove,removeUnusedMets,removeUnusedGenes,removeUnusedComps)">removeReactions</a>(models{i},hasNoGenes,true,true);
0168 <span class="keyword">end</span>
0169 
0170 <span class="comment">%Create a structure that contains all genes used in the blasts in any</span>
0171 <span class="comment">%direction for each of the models in 'models' and for the new organism. The</span>
0172 <span class="comment">%first cell is for the new organism and then according to the preferred</span>
0173 <span class="comment">%order. If no such order is supplied, then according to the order in</span>
0174 <span class="comment">%'models'</span>
0175 allGenes=cell(numel(models)+1,1);
0176 <span class="keyword">if</span> isempty(preferredOrder)
0177     useOrder=modelNames;
0178 <span class="keyword">else</span>
0179     useOrder=preferredOrder;
0180 <span class="keyword">end</span>
0181 
0182 <span class="comment">%Get the corresponding indexes for those models in the 'models' structure</span>
0183 useOrderIndexes=zeros(numel(models),1);
0184 <span class="keyword">for</span> i=1:numel(models)
0185     [~, index]=ismember(models{i}.id,useOrder);
0186     useOrderIndexes(index)=i;
0187 <span class="keyword">end</span>
0188 
0189 <span class="comment">%Remove all genes from the blast structure that have no genes in the models</span>
0190 <span class="keyword">if</span> onlyGenesInModels==true
0191     modelGenes={};
0192     <span class="keyword">for</span> i=1:numel(models)
0193         modelGenes=[modelGenes;models{i}.genes];
0194     <span class="keyword">end</span>
0195     <span class="keyword">for</span> i=1:numel(blastStructure)
0196         <span class="comment">%Check to see if it should match the toId or fromId</span>
0197         <span class="keyword">if</span> strcmpi(blastStructure(i).fromId,getModelFor)
0198             I=ismember(blastStructure(i).toGenes,modelGenes);
0199         <span class="keyword">else</span>
0200             I=ismember(blastStructure(i).fromGenes,modelGenes);
0201         <span class="keyword">end</span>
0202         blastStructure(i).fromGenes(~I)=[];
0203         blastStructure(i).toGenes(~I)=[];
0204         blastStructure(i).evalue(~I)=[];
0205         blastStructure(i).identity(~I)=[];
0206         blastStructure(i).aligLen(~I)=[];
0207         blastStructure(i).bitscore(~I)=[];
0208         blastStructure(i).ppos(~I)=[];
0209         
0210         <span class="comment">%Check that no matching in blastStructure is empty. This happens if</span>
0211         <span class="comment">%no genes in the models are present in the corresponding sheet</span>
0212         <span class="keyword">if</span> isempty(blastStructure(i).fromGenes)
0213             EM=[<span class="string">'No genes in matching from '</span> blastStructure(i).fromId <span class="string">' to '</span> blastStructure(i).toId <span class="string">' are present in the corresponding model'</span>];
0214             <a href="dispEM.html" class="code" title="function dispEM(string,throwErrors,toList,trimWarnings)">dispEM</a>(EM);
0215         <span class="keyword">end</span>
0216     <span class="keyword">end</span>
0217 <span class="keyword">end</span>
0218 
0219 <span class="comment">%If only best orthologs are to be used then all other measurements are</span>
0220 <span class="comment">%deleted from the blastStructure. All code after this stays the same. This</span>
0221 <span class="comment">%means that preferred order can still matter. The best ortholog scoring is</span>
0222 <span class="comment">%based only on the E-value</span>
0223 <span class="keyword">if</span> strictness==3
0224     <span class="keyword">for</span> i=1:numel(blastStructure)
0225         keep=false(numel(blastStructure(i).toGenes),1);
0226         [allFromGenes, ~, I]=unique(blastStructure(i).fromGenes);
0227         
0228         <span class="comment">%It would be nice to get rid of this loop</span>
0229         <span class="keyword">for</span> j=1:numel(allFromGenes)
0230             allMatches=find(I==j);
0231             bestMatches=allMatches(blastStructure(i).evalue(allMatches)==min(blastStructure(i).evalue(allMatches)));
0232             
0233             <span class="comment">%Keep the best matches</span>
0234             keep(bestMatches)=true;
0235         <span class="keyword">end</span>
0236         
0237         <span class="comment">%Delete all matches that were not best matches</span>
0238         blastStructure(i).fromGenes(~keep)=[];
0239         blastStructure(i).toGenes(~keep)=[];
0240         blastStructure(i).evalue(~keep)=[];
0241         blastStructure(i).identity(~keep)=[];
0242         blastStructure(i).aligLen(~keep)=[];
0243         blastStructure(i).bitscore(~keep)=[];
0244         blastStructure(i).ppos(~keep)=[];
0245     <span class="keyword">end</span>
0246 <span class="keyword">end</span>
0247 
0248 useOrder=[{getModelFor};useOrder];
0249 
0250 <span class="keyword">for</span> i=1:numel(blastStructure)
0251     [~, toIndex]=ismember(blastStructure(i).toId,useOrder);
0252     [~, fromIndex]=ismember(blastStructure(i).fromId,useOrder);
0253     
0254     <span class="comment">%Add all genes to the corresponding list in allGenes</span>
0255     allGenes{toIndex}=[allGenes{toIndex};blastStructure(i).toGenes];
0256     allGenes{fromIndex}=[allGenes{fromIndex};blastStructure(i).fromGenes];
0257 <span class="keyword">end</span>
0258 
0259 <span class="comment">%Keep only the unique gene names</span>
0260 maxOtherGeneNr=0; <span class="comment">%Determines the dimension of the connectivity matrixes</span>
0261 <span class="keyword">for</span> i=1:numel(allGenes)
0262     allGenes{i}=unique(allGenes{i});
0263     <span class="keyword">if</span> i&gt;1
0264         <span class="keyword">if</span> numel(allGenes{i})&gt;maxOtherGeneNr
0265             maxOtherGeneNr=numel(allGenes{i});
0266         <span class="keyword">end</span>
0267     <span class="keyword">end</span>
0268 <span class="keyword">end</span>
0269 
0270 <span class="comment">%Generate a cell array of matrixes that describes how the genes in the new</span>
0271 <span class="comment">%organism map to the models. Each cell matches to the corresponding model</span>
0272 <span class="comment">%in useOrder (starting at 2 of course). First dimension is gene in new</span>
0273 <span class="comment">%organism, second which gene it is in the other organism. The second matrix</span>
0274 <span class="comment">%describes how they map back.</span>
0275 
0276 <span class="comment">%As it is now, a significant match is indicated by a 1. This could be</span>
0277 <span class="comment">%expanded to contain some kind of significance level. The first dimension</span>
0278 <span class="comment">%is still the genes in the new model.</span>
0279 
0280 allTo=cell(numel(useOrder)-1,1);
0281 allFrom=cell(numel(useOrder)-1,1);
0282 
0283 <span class="keyword">for</span> i=1:numel(useOrder)-1
0284     allTo{i}=sparse(numel(allGenes{1}),numel(allGenes{i+1}));
0285     allFrom{i}=sparse(numel(allGenes{1}),numel(allGenes{i+1}));
0286 <span class="keyword">end</span>
0287 
0288 <span class="comment">%Fill the matches to other species</span>
0289 <span class="keyword">for</span> i=1:numel(blastStructure)
0290     <span class="keyword">if</span> strcmp(blastStructure(i).toId,getModelFor)
0291         <span class="comment">%This was 'to' the new organism. They should all match so no checks</span>
0292         <span class="comment">%are being made</span>
0293         [~, a]=ismember(blastStructure(i).toGenes,allGenes{1});
0294         [~, fromModel]=ismember(blastStructure(i).fromId,useOrder);
0295         [~, b]=ismember(blastStructure(i).fromGenes,allGenes{fromModel});
0296         idx = sub2ind(size(allTo{fromModel-1}), a, b);
0297         allTo{fromModel-1}(idx)=1;
0298     <span class="keyword">else</span>
0299         <span class="comment">%This was 'from' the new organism</span>
0300         [~, a]=ismember(blastStructure(i).fromGenes,allGenes{1});
0301         [~, toModel]=ismember(blastStructure(i).toId,useOrder);
0302         [~, b]=ismember(blastStructure(i).toGenes,allGenes{toModel});
0303         idx = sub2ind(size(allFrom{toModel-1}), a, b);
0304         allFrom{toModel-1}(idx)=1;
0305     <span class="keyword">end</span>
0306 <span class="keyword">end</span>
0307 
0308 <span class="comment">%Now we have all the gene matches in a convenient way. For all the genes in</span>
0309 <span class="comment">%the new organism get the genes that should be included from other</span>
0310 <span class="comment">%organisms. If all genes should be included this simply means keep the</span>
0311 <span class="comment">%allFrom matrix as it is. If only orthologs which could be mapped in both</span>
0312 <span class="comment">%BLAST directions are to be included then only those elements are kept.</span>
0313 
0314 finalMappings=cell(numel(useOrder)-1,1);
0315 <span class="keyword">if</span> strictness==1 || strictness==3
0316     <span class="keyword">for</span> j=1:numel(allFrom)
0317         finalMappings{j}=allTo{j}~=0 &amp; allFrom{j}~=0;
0318     <span class="keyword">end</span>
0319 <span class="keyword">else</span>
0320     <span class="keyword">if</span> mapNewGenesToOld==true
0321         finalMappings=allFrom;
0322     <span class="keyword">else</span>
0323         finalMappings=allTo;
0324     <span class="keyword">end</span>
0325 <span class="keyword">end</span>
0326 
0327 <span class="comment">%Remove all genes from the mapping that are not in the models. This doesn't</span>
0328 <span class="comment">%do much if only genes in the models were used for the original mapping.</span>
0329 <span class="comment">%Also simplify the finalMapping and allGenes structures so that they only</span>
0330 <span class="comment">%contain mappings that exist</span>
0331 usedNewGenes=false(numel(allGenes{1}),1);
0332 
0333 <span class="keyword">for</span> i=1:numel(allGenes)-1
0334     <span class="comment">%First remove mappings for those genes that are not in the model</span>
0335     <span class="keyword">if</span> onlyGenesInModels==false
0336         a=ismember(allGenes{i+1},models{useOrderIndexes(i)}.genes);
0337         finalMappings{i}(:,~a)=false;
0338     <span class="keyword">end</span>
0339     
0340     <span class="comment">%Then remove unused ones and simplify</span>
0341     [a, b]=find(finalMappings{i});
0342     usedGenes=false(numel(allGenes{i+1}),1);
0343     usedNewGenes(a)=true;
0344     usedGenes(b)=true;
0345     finalMappings{i}=finalMappings{i}(:,usedGenes);
0346     allGenes{i+1}=allGenes{i+1}(usedGenes);
0347 <span class="keyword">end</span>
0348 
0349 <span class="comment">%Remove all new genes that have not been mapped to anything</span>
0350 allGenes{1}=allGenes{1}(usedNewGenes);
0351 <span class="keyword">for</span> i=1:numel(finalMappings)
0352     finalMappings{i}=finalMappings{i}(usedNewGenes,:);
0353 <span class="keyword">end</span>
0354 
0355 <span class="comment">%Now is it time to choose which reactions should be included from which</span>
0356 <span class="comment">%models. If there is a preferred order specified then each gene can only</span>
0357 <span class="comment">%result in reactions from one model, otherwise they should all be included</span>
0358 
0359 <span class="comment">%Start by simplifying the models by removing genes/reactions that are not</span>
0360 <span class="comment">%used. This is where it gets weird with complexes, especially &quot;or&quot;</span>
0361 <span class="comment">%complexes. In this step only reactions which are encoded by one single</span>
0362 <span class="comment">%gene, or where all genes should be deleted, are deleted. The info on the</span>
0363 <span class="comment">%full complex is still present in the grRules</span>
0364 
0365 <span class="keyword">for</span> i=1:numel(models)
0366     a=ismember(models{useOrderIndexes(i)}.genes,allGenes{i+1});
0367     
0368     <span class="comment">%Remove reactions that are not associated to any of the genes in</span>
0369     <span class="comment">%allGenes, thereby also keeping complexes where only for one of the</span>
0370     <span class="comment">%genes was matched</span>
0371     [rxnsToKeep,~] = find(models{useOrderIndexes(i)}.rxnGeneMat(:,a));
0372     rxnsToRemove = repmat(1,numel(models{useOrderIndexes(i)}.rxns),1);
0373     rxnsToRemove(rxnsToKeep) = 0;
0374     rxnsToRemove = find(rxnsToRemove);
0375     models{useOrderIndexes(i)}=<a href="removeReactions.html" class="code" title="function reducedModel=removeReactions(model,rxnsToRemove,removeUnusedMets,removeUnusedGenes,removeUnusedComps)">removeReactions</a>(models{useOrderIndexes(i)},rxnsToRemove,true,true,true);
0376 <span class="keyword">end</span>
0377 
0378 <span class="comment">%Since mergeModels function will be used in the end, the models are</span>
0379 <span class="comment">%simplified further by deleting genes/reactions in the order specified by</span>
0380 <span class="comment">%preferredOrder. This means that the last model will only contain reactions</span>
0381 <span class="comment">%for genes that mapped only to that model</span>
0382 
0383 allUsedGenes=false(numel(allGenes{1}),1);
0384 
0385 <span class="keyword">if</span> ~isempty(preferredOrder) &amp;&amp; numel(models)&gt;1
0386     [usedGenes, ~]=find(finalMappings{1}); <span class="comment">%All that are used in the first model in preferredOrder</span>
0387     allUsedGenes(usedGenes)=true;
0388     <span class="keyword">for</span> i=2:numel(finalMappings)
0389         [usedGenes, ~]=find(finalMappings{i});
0390         usedGenes=unique(usedGenes);
0391         a=ismember(usedGenes,find(allUsedGenes));
0392         
0393         [~, genesToDelete]=find(finalMappings{i}(usedGenes(a),:)); <span class="comment">%IMPORTANT! IS it really correct to remove all genes that map?</span>
0394         genesToDelete=unique(genesToDelete); <span class="comment">%Maybe not needed, but for clarity if nothing else</span>
0395         
0396         <span class="comment">%Remove all the genes that were already found and add the other</span>
0397         <span class="comment">%ones to allUsedGenes</span>
0398         models{useOrderIndexes(i)}=<a href="removeGenes.html" class="code" title="function reducedModel = removeGenes(model,genesToRemove,removeUnusedMets,removeBlockedRxns,standardizeRules)">removeGenes</a>(models{useOrderIndexes(i)},allGenes{i+1}(genesToDelete),true,true,false);
0399         allUsedGenes(usedGenes)=true;
0400         
0401         <span class="comment">%Remove the deleted genes from finalMappings and allGenes.</span>
0402         finalMappings{i}(:,genesToDelete)=[];
0403         allGenes{i+1}(genesToDelete)=[];
0404     <span class="keyword">end</span>
0405 <span class="keyword">end</span>
0406 
0407 <span class="comment">%Now loop through the models and update the gene associations. Genes not</span>
0408 <span class="comment">%belonging to the new organism will be renamed as 'OLD_MODELID_gene'</span>
0409 <span class="keyword">for</span> i=1:numel(models)
0410     <span class="comment">%Find all the new genes that should be used for this model</span>
0411     [newGenes, oldGenes]=find(finalMappings{i});
0412     
0413     <span class="comment">%Create a new gene list with the genes from the new organism and those</span>
0414     <span class="comment">%genes that could not be removed</span>
0415     replaceableGenes=allGenes{i+1}(unique(oldGenes));
0416     nonReplaceableGenes=setdiff(models{useOrderIndexes(i)}.genes,replaceableGenes);
0417     fullGeneList=[allGenes{1}(unique(newGenes));nonReplaceableGenes];
0418     
0419     <span class="comment">%Just to save some indexing later. This is the LAST index of</span>
0420     <span class="comment">%replaceable ones</span>
0421     nonRepStartIndex=numel(unique(newGenes));
0422     
0423     <span class="comment">%Construct a new rxnGeneMat</span>
0424     newRxnGeneMat=sparse(numel(models{useOrderIndexes(i)}.rxns),numel(fullGeneList));
0425     
0426     <span class="comment">%Now update the rxnGeneMat. This is a little tricky and could</span>
0427     <span class="comment">%probably be done in a more efficient way, but I just loop through the</span>
0428     <span class="comment">%reactions and add them one by one</span>
0429     <span class="keyword">for</span> j=1:numel(models{useOrderIndexes(i)}.rxns)
0430         <span class="comment">%Get the old genes encoding for this reaction</span>
0431         [~, oldGeneIds]=find(models{useOrderIndexes(i)}.rxnGeneMat(j,:));
0432         
0433         <span class="comment">%Update the matrix for each gene. This includes replacing one gene</span>
0434         <span class="comment">%with several new ones if there were several matches</span>
0435         <span class="keyword">for</span> k=1:numel(oldGeneIds)
0436             <span class="comment">%Match the gene to one in the gene list. This is done as a text</span>
0437             <span class="comment">%match. Could probably be done better, but I'm a little lost in</span>
0438             <span class="comment">%the indexing</span>
0439             
0440             geneName=models{useOrderIndexes(i)}.genes(oldGeneIds(k));
0441             
0442             <span class="comment">%First search in the mappable genes</span>
0443             mapIndex=find(ismember(allGenes{i+1},geneName));
0444             
0445             <span class="keyword">if</span> ~isempty(mapIndex)
0446                 <span class="comment">% add the old genes</span>
0447                 hitGenes.oldGenes = [hitGenes.oldGenes, {geneName}];
0448                 
0449                 <span class="comment">%Get the new genes for that gene</span>
0450                 a=find(finalMappings{i}(:,mapIndex));
0451                 
0452                 <span class="comment">%Find the positions of these genes in the final gene list</span>
0453                 [~, b]=ismember(allGenes{1}(a),fullGeneList);
0454                 
0455                 <span class="comment">%Update the matrix</span>
0456                 newRxnGeneMat(j,b)=1;
0457                 
0458                 <span class="comment">%Update the grRules string. This is tricky, but I hope that</span>
0459                 <span class="comment">%it's ok to replace the old gene name with the new one and</span>
0460                 <span class="comment">%add ') or (' if there were several matches. Be sure of</span>
0461                 <span class="comment">%this!</span>
0462                 repString=fullGeneList{b(1)};
0463                 <span class="keyword">if</span> numel(b)&gt;1
0464                     <span class="keyword">for</span> l=2:numel(b)
0465                         repString=[repString <span class="string">' or '</span> fullGeneList{b(l)}];
0466                     <span class="keyword">end</span>
0467                     repString=[<span class="string">'('</span> repString <span class="string">')'</span>];
0468                 <span class="keyword">end</span>
0469                 
0470                 <span class="comment">% add the new matched genes</span>
0471                 hitGenes.newGenes = [hitGenes.newGenes, {repString}];
0472                 
0473                 <span class="comment">%Use regexprep instead of strrep to prevent partial matches</span>
0474                 models{useOrderIndexes(i)}.grRules{j}=regexprep(models{useOrderIndexes(i)}.grRules{j},[<span class="string">'(^|\s|\()'</span> geneName{1} <span class="string">'($|\s|\))'</span>],[<span class="string">'$1'</span> repString <span class="string">'$2'</span>]);
0475             <span class="keyword">else</span>
0476                 <span class="comment">%Then search in the non-replaceable genes. There could only</span>
0477                 <span class="comment">%be one match here</span>
0478                 index=find(ismember(nonReplaceableGenes,geneName));
0479                 
0480                 <span class="comment">%Update the matrix</span>
0481                 newRxnGeneMat(j,nonRepStartIndex+index)=1;
0482                 
0483                 models{useOrderIndexes(i)}.grRules{j}=strrep(models{useOrderIndexes(i)}.grRules{j},geneName{1},strcat(<span class="string">'OLD_'</span>,models{useOrderIndexes(i)}.id,<span class="string">'_'</span>,geneName{1}));
0484             <span class="keyword">end</span>
0485         <span class="keyword">end</span>
0486     <span class="keyword">end</span>
0487     
0488     <span class="comment">%Add the new list of genes</span>
0489     models{useOrderIndexes(i)}.rxnGeneMat=newRxnGeneMat;
0490     <span class="keyword">if</span> ~isempty(nonReplaceableGenes)
0491         models{useOrderIndexes(i)}.genes=[allGenes{1}(unique(newGenes));strcat(<span class="string">'OLD_'</span>,models{useOrderIndexes(i)}.id,<span class="string">'_'</span>,nonReplaceableGenes)];
0492     <span class="keyword">else</span>
0493         models{useOrderIndexes(i)}.genes=allGenes{1}(unique(newGenes));
0494     <span class="keyword">end</span>
0495     <span class="keyword">if</span> isfield(models{useOrderIndexes(i)},<span class="string">'geneComps'</span>)
0496         geneComps=models{useOrderIndexes(i)}.geneComps(1);
0497         models{useOrderIndexes(i)}.geneComps=zeros(numel(models{useOrderIndexes(i)}.genes),1);
0498         <span class="comment">%Assume that all genes are in the same compartment, and this</span>
0499         <span class="comment">%compartment is specified for the first gene</span>
0500         models{useOrderIndexes(i)}.geneComps(:)=geneComps;
0501     <span class="keyword">end</span>
0502 <span class="keyword">end</span>
0503 
0504 <span class="comment">%Now merge the models. All information should be correct except for 'or'</span>
0505 <span class="comment">%complexes</span>
0506 draftModel=<a href="mergeModels.html" class="code" title="function model=mergeModels(models,metParam,supressWarnings)">mergeModels</a>(models,<span class="string">'metNames'</span>);
0507 
0508 <span class="comment">%Change name of the resulting model</span>
0509 draftModel.id=getModelFor;
0510 name=<span class="string">'Generated by getModelFromHomology using '</span>;
0511 <span class="keyword">for</span> i=1:numel(models)
0512     <span class="keyword">if</span> i&lt;numel(models)
0513         name=[name models{i}.id <span class="string">', '</span>];
0514     <span class="keyword">else</span>
0515         name=[name models{i}.id];
0516     <span class="keyword">end</span>
0517 <span class="keyword">end</span>
0518 draftModel.name=name;
0519 draftModel.rxnNotes=cell(length(draftModel.rxns),1);
0520 draftModel.rxnNotes(:)={<span class="string">'Included by getModelFromHomology'</span>};
0521 draftModel.rxnConfidenceScores=NaN(length(draftModel.rxns),1);
0522 draftModel.rxnConfidenceScores(:)=2;
0523 draftModel=<a href="deleteUnusedGenes.html" class="code" title="function reducedModel=deleteUnusedGenes(model,verbose)">deleteUnusedGenes</a>(draftModel,0);
0524 <span class="comment">%Standardize grRules and notify if problematic grRules are found</span>
0525 [draftModel.grRules,draftModel.rxnGeneMat]=<a href="standardizeGrRules.html" class="code" title="function [grRules,rxnGeneMat,indexes2check] = standardizeGrRules(model,embedded)">standardizeGrRules</a>(draftModel,false);
0526 <span class="keyword">end</span></pre></div>
<hr><address>Generated by <strong><a href="http://www.artefact.tk/software/matlab/m2html/" title="Matlab Documentation in HTML">m2html</a></strong> &copy; 2005</address>
</body>
</html>