<!DOCTYPE HTML PUBLIC "-//W3C//DTD HTML 4.01 Transitional//EN"
                "http://www.w3.org/TR/REC-html40/loose.dtd">
<html>
<head>
  <title>Description of getModelFromHomology</title>
  <meta name="keywords" content="getModelFromHomology">
  <meta name="description" content="getModelFromHomology">
  <meta http-equiv="Content-Type" content="text/html; charset=iso-8859-1">
  <meta name="generator" content="m2html v1.5 &copy; 2003-2005 Guillaume Flandin">
  <meta name="robots" content="index, follow">
  <link type="text/css" rel="stylesheet" href="../m2html.css">
</head>
<body>
<a name="_top"></a>
<div><a href="../index.html">Home</a> &gt;  <a href="index.html">core</a> &gt; getModelFromHomology.m</div>

<!--<table width="100%"><tr><td align="left"><a href="../index.html"><img alt="<" border="0" src="../left.png">&nbsp;Master index</a></td>
<td align="right"><a href="index.html">Index for core&nbsp;<img alt=">" border="0" src="../right.png"></a></td></tr></table>-->

<h1>getModelFromHomology
</h1>

<h2><a name="_name"></a>PURPOSE <a href="#_top"><img alt="^" border="0" src="../up.png"></a></h2>
<div class="box"><strong>getModelFromHomology</strong></div>

<h2><a name="_synopsis"></a>SYNOPSIS <a href="#_top"><img alt="^" border="0" src="../up.png"></a></h2>
<div class="box"><strong>function [draftModel, hitGenes]=getModelFromHomology(models,blastStructure,getModelFor,preferredOrder,strictness,onlyGenesInModels,maxE,minLen,minIde,mapNewGenesToOld) </strong></div>

<h2><a name="_description"></a>DESCRIPTION <a href="#_top"><img alt="^" border="0" src="../up.png"></a></h2>
<div class="fragment"><pre class="comment"> getModelFromHomology
   Constructs a new model from a set of existing models and gene homology
   information.

   models            a cell array of model structures to build the model
                     from. These models must be sorted by importance in
                     decreasing order
   blastStructure    a blastStructure as produced by getBlast or
                     getBlastFromExcel
   getModelFor       a three-four letter abbreviation of the organism to
                     build a model for. Must have BLASTP hits in both
                     directions to the organisms in 'models'
   preferredOrder    the order in which reactions should be added from the
                     models. If not supplied, reactions will be included
                     from all models, otherwise one gene will only result
                     in reactions from one model (opt, default {})
   strictness        integer that specifies which reactions should be
                     included:
                     1: Map new genes to old for all pairs, which have
                     acceptable BLASTP results in both directions
                     2: Map new genes to old for all pairs, which have
                     acceptable BLASTP results in correspondent direction
                     (mapping can be done in the opposite direction, see
                     mapNewGenesToOld below)
                     3: Check all BLASTP results and retain only the best
                     results by E-value for all gene pairs in each
                     direction separately. Then map new genes to old for
                     all pairs, which have acceptable BLASTP results in
                     both directions (opt, default 1).
   onlyGenesInModels consider BLASTP results only for genes that exist in
                     the models. This tends to import a larger fraction
                     from the existing models but may give less reliable
                     results. Has effect only if strictness=3 (opt,
                     default false)
   maxE              only look at genes with E-values &lt;= this value (opt,
                     default 10^-30)
   minLen            only look at genes with alignment length &gt;= this
                     value (opt, default 200)
   minIde            only look at genes with identity &gt;= this value
                     (opt, default 40 (%))
   mapNewGenesToOld  determines how to match genes if not looking at only
                     1-1 orthologs. Either map the new genes to the old or
                     old genes to new. The default is to map the new genes
                     (opt, default true)

   draftModel        a model structure for the new organism
   hitGenes          collect the old and new genes

   The models in the 'models' structure should have named the metabolites
   in the same manner, have their reversible reactions in the same
   direction (run sortModel), and use the same compartment names. To avoid
   keeping unneccesary old genes, the models should not have
   'or'-relations in their grRules (use expandModel).

   The resulting draft model contains only reactions associated with
   orthologous genes. The old (original) genes involved in 'and'
   relations in grRules without any orthologs are still included in
   the draft model as OLD_MODELID_geneName.

   NOTE: &quot;to&quot; and &quot;from&quot; means relative to the new organism

   Usage: draftModel=getModelFromHomology(models,blastStructure,...
    getModelFor,preferredOrder,strictness,onlyGenesInModels,maxE,...
    minLen,minIde,mapNewGenesToOld)</pre></div>

<!-- crossreference -->
<h2><a name="_cross"></a>CROSS-REFERENCE INFORMATION <a href="#_top"><img alt="^" border="0" src="../up.png"></a></h2>
This function calls:
<ul style="list-style-image:url(../matlabicon.gif)">
<li><a href="convertCharArray.html" class="code" title="function inputConverted = convertCharArray(funcInput)">convertCharArray</a>	convertCharArray</li><li><a href="deleteUnusedGenes.html" class="code" title="function reducedModel=deleteUnusedGenes(model,verbose)">deleteUnusedGenes</a>	deleteUnusedGenes</li><li><a href="dispEM.html" class="code" title="function dispEM(string,throwErrors,toList,trimWarnings)">dispEM</a>	dispEM</li><li><a href="mergeModels.html" class="code" title="function model=mergeModels(models,metParam,supressWarnings)">mergeModels</a>	mergeModels</li><li><a href="removeGenes.html" class="code" title="function reducedModel = removeGenes(model,genesToRemove,removeUnusedMets,removeBlockedRxns,standardizeRules)">removeGenes</a>	removeGenes</li><li><a href="removeReactions.html" class="code" title="function reducedModel=removeReactions(model,rxnsToRemove,removeUnusedMets,removeUnusedGenes,removeUnusedComps)">removeReactions</a>	removeReactions</li><li><a href="standardizeGrRules.html" class="code" title="function [grRules,rxnGeneMat,indexes2check] = standardizeGrRules(model,embedded)">standardizeGrRules</a>	standardizeGrRules</li></ul>
This function is called by:
<ul style="list-style-image:url(../matlabicon.gif)">
</ul>
<!-- crossreference -->



<h2><a name="_source"></a>SOURCE CODE <a href="#_top"><img alt="^" border="0" src="../up.png"></a></h2>
<div class="fragment"><pre>0001 <a name="_sub0" href="#_subfunctions" class="code">function [draftModel, hitGenes]=getModelFromHomology(models,blastStructure,</a><span class="keyword">...</span>
0002     getModelFor,preferredOrder,strictness,onlyGenesInModels,maxE,<span class="keyword">...</span>
0003     minLen,minIde,mapNewGenesToOld)
0004 <span class="comment">% getModelFromHomology</span>
0005 <span class="comment">%   Constructs a new model from a set of existing models and gene homology</span>
0006 <span class="comment">%   information.</span>
0007 <span class="comment">%</span>
0008 <span class="comment">%   models            a cell array of model structures to build the model</span>
0009 <span class="comment">%                     from. These models must be sorted by importance in</span>
0010 <span class="comment">%                     decreasing order</span>
0011 <span class="comment">%   blastStructure    a blastStructure as produced by getBlast or</span>
0012 <span class="comment">%                     getBlastFromExcel</span>
0013 <span class="comment">%   getModelFor       a three-four letter abbreviation of the organism to</span>
0014 <span class="comment">%                     build a model for. Must have BLASTP hits in both</span>
0015 <span class="comment">%                     directions to the organisms in 'models'</span>
0016 <span class="comment">%   preferredOrder    the order in which reactions should be added from the</span>
0017 <span class="comment">%                     models. If not supplied, reactions will be included</span>
0018 <span class="comment">%                     from all models, otherwise one gene will only result</span>
0019 <span class="comment">%                     in reactions from one model (opt, default {})</span>
0020 <span class="comment">%   strictness        integer that specifies which reactions should be</span>
0021 <span class="comment">%                     included:</span>
0022 <span class="comment">%                     1: Map new genes to old for all pairs, which have</span>
0023 <span class="comment">%                     acceptable BLASTP results in both directions</span>
0024 <span class="comment">%                     2: Map new genes to old for all pairs, which have</span>
0025 <span class="comment">%                     acceptable BLASTP results in correspondent direction</span>
0026 <span class="comment">%                     (mapping can be done in the opposite direction, see</span>
0027 <span class="comment">%                     mapNewGenesToOld below)</span>
0028 <span class="comment">%                     3: Check all BLASTP results and retain only the best</span>
0029 <span class="comment">%                     results by E-value for all gene pairs in each</span>
0030 <span class="comment">%                     direction separately. Then map new genes to old for</span>
0031 <span class="comment">%                     all pairs, which have acceptable BLASTP results in</span>
0032 <span class="comment">%                     both directions (opt, default 1).</span>
0033 <span class="comment">%   onlyGenesInModels consider BLASTP results only for genes that exist in</span>
0034 <span class="comment">%                     the models. This tends to import a larger fraction</span>
0035 <span class="comment">%                     from the existing models but may give less reliable</span>
0036 <span class="comment">%                     results. Has effect only if strictness=3 (opt,</span>
0037 <span class="comment">%                     default false)</span>
0038 <span class="comment">%   maxE              only look at genes with E-values &lt;= this value (opt,</span>
0039 <span class="comment">%                     default 10^-30)</span>
0040 <span class="comment">%   minLen            only look at genes with alignment length &gt;= this</span>
0041 <span class="comment">%                     value (opt, default 200)</span>
0042 <span class="comment">%   minIde            only look at genes with identity &gt;= this value</span>
0043 <span class="comment">%                     (opt, default 40 (%))</span>
0044 <span class="comment">%   mapNewGenesToOld  determines how to match genes if not looking at only</span>
0045 <span class="comment">%                     1-1 orthologs. Either map the new genes to the old or</span>
0046 <span class="comment">%                     old genes to new. The default is to map the new genes</span>
0047 <span class="comment">%                     (opt, default true)</span>
0048 <span class="comment">%</span>
0049 <span class="comment">%   draftModel        a model structure for the new organism</span>
0050 <span class="comment">%   hitGenes          collect the old and new genes</span>
0051 <span class="comment">%</span>
0052 <span class="comment">%   The models in the 'models' structure should have named the metabolites</span>
0053 <span class="comment">%   in the same manner, have their reversible reactions in the same</span>
0054 <span class="comment">%   direction (run sortModel), and use the same compartment names. To avoid</span>
0055 <span class="comment">%   keeping unneccesary old genes, the models should not have</span>
0056 <span class="comment">%   'or'-relations in their grRules (use expandModel).</span>
0057 <span class="comment">%</span>
0058 <span class="comment">%   The resulting draft model contains only reactions associated with</span>
0059 <span class="comment">%   orthologous genes. The old (original) genes involved in 'and'</span>
0060 <span class="comment">%   relations in grRules without any orthologs are still included in</span>
0061 <span class="comment">%   the draft model as OLD_MODELID_geneName.</span>
0062 <span class="comment">%</span>
0063 <span class="comment">%   NOTE: &quot;to&quot; and &quot;from&quot; means relative to the new organism</span>
0064 <span class="comment">%</span>
0065 <span class="comment">%   Usage: draftModel=getModelFromHomology(models,blastStructure,...</span>
0066 <span class="comment">%    getModelFor,preferredOrder,strictness,onlyGenesInModels,maxE,...</span>
0067 <span class="comment">%    minLen,minIde,mapNewGenesToOld)</span>
0068 
0069 hitGenes.oldGenes = [];  <span class="comment">% collect the old genes from the template model (organism)</span>
0070 hitGenes.newGenes = [];  <span class="comment">% collect the new genes of the draft model (target organism)</span>
0071 
0072 getModelFor=char(getModelFor);
0073 
0074 <span class="keyword">if</span> nargin&lt;4
0075     preferredOrder=[];
0076 <span class="keyword">else</span>
0077     preferredOrder=<a href="convertCharArray.html" class="code" title="function inputConverted = convertCharArray(funcInput)">convertCharArray</a>(preferredOrder);
0078     [row,col]=size(preferredOrder);
0079     <span class="keyword">if</span> col&gt;row
0080         preferredOrder=transpose(preferredOrder);
0081     <span class="keyword">end</span>
0082 <span class="keyword">end</span>
0083 <span class="keyword">if</span> nargin&lt;5
0084     strictness=1;
0085 <span class="keyword">end</span>
0086 <span class="keyword">if</span> nargin&lt;6
0087     onlyGenesInModels=false;
0088 <span class="keyword">end</span>
0089 <span class="keyword">if</span> nargin&lt;7
0090     maxE=10^-30;
0091 <span class="keyword">end</span>
0092 <span class="keyword">if</span> nargin&lt;8
0093     minLen=200;
0094 <span class="keyword">end</span>
0095 <span class="keyword">if</span> nargin&lt;9
0096     minIde=40;
0097 <span class="keyword">end</span>
0098 <span class="keyword">if</span> nargin&lt;10
0099     mapNewGenesToOld=true;
0100 <span class="keyword">end</span>
0101 
0102 <span class="keyword">if</span> isfield(models,<span class="string">'S'</span>)
0103     models={models};
0104 <span class="keyword">end</span>
0105     
0106 <span class="comment">%Check that all the information is in the blast structure</span>
0107 modelNames=cell(numel(models),1);
0108 <span class="keyword">for</span> i=1:numel(models)
0109     modelNames{i}=models{i}.id;
0110     <span class="comment">%Gene short names and geneMiriams are often different between species,</span>
0111     <span class="comment">%safer not to include them</span>
0112     <span class="keyword">if</span> isfield(models{i},<span class="string">'geneShortNames'</span>)
0113         models{i}=rmfield(models{i},<span class="string">'geneShortNames'</span>);
0114     <span class="keyword">end</span>
0115     <span class="keyword">if</span> isfield(models{i},<span class="string">'geneMiriams'</span>)
0116         models{i}=rmfield(models{i},<span class="string">'geneMiriams'</span>);
0117     <span class="keyword">end</span>
0118 <span class="keyword">end</span>
0119 
0120 <span class="comment">%Assume for now that all information is there and that it's correct. This</span>
0121 <span class="comment">%is important to fix since no further checks are being made!</span>
0122 
0123 <span class="comment">%Check whether provided fasta files use the same gene identifiers as</span>
0124 <span class="comment">%provided template models</span>
0125 <span class="keyword">for</span> i=1:numel(blastStructure)
0126     <span class="keyword">if</span> ~strcmp(blastStructure(i).fromId,getModelFor)
0127         j=strcmpi(blastStructure(i).fromId,modelNames);
0128         <span class="keyword">if</span> j==0
0129             error([<span class="string">'While the blastStructure contains sequences from '</span><span class="keyword">...</span>
0130                 <span class="string">'organismID &quot;%s&quot; (as\nprovided in getBlast), none of '</span><span class="keyword">...</span>
0131                 <span class="string">'template models have this id (as model.id)'</span>],<span class="keyword">...</span>
0132                 string(blastStructure(i).fromId));
0133         <span class="keyword">end</span>
0134         k=sum(ismember(blastStructure(i).fromGenes,models{j}.genes));
0135         <span class="keyword">if</span> k&lt;(numel(models{j}.genes)*0.05)
0136             error([<span class="string">'Less than 5%% of the genes in the template model '</span><span class="keyword">...</span>
0137                 <span class="string">'with model.id &quot;%s&quot;\ncan be found in the blastStructure. '</span><span class="keyword">...</span>
0138                 <span class="string">'Ensure that the protein FASTA\nused in getBlast and '</span><span class="keyword">...</span>
0139                 <span class="string">'the template model used in getModelFromHomology\nuse '</span><span class="keyword">...</span>
0140                 <span class="string">'the same style of gene identifiers'</span>],models{j}.id)
0141         <span class="keyword">end</span>
0142     <span class="keyword">end</span>
0143 <span class="keyword">end</span>
0144 
0145 <span class="comment">%Standardize grRules of template models</span>
0146 <span class="keyword">for</span> i=1:length(models)
0147     fprintf(<span class="string">'\nStandardizing grRules of template model with ID &quot;%s&quot; ...'</span>,models{i}.id);
0148     [models{i}.grRules,models{i}.rxnGeneMat]=<a href="standardizeGrRules.html" class="code" title="function [grRules,rxnGeneMat,indexes2check] = standardizeGrRules(model,embedded)">standardizeGrRules</a>(models{i},false);
0149 <span class="keyword">end</span>
0150 fprintf(<span class="string">' done\n'</span>);
0151 
0152 <span class="comment">%Remove all gene matches that are below the cutoffs</span>
0153 <span class="keyword">for</span> i=1:numel(blastStructure)
0154     indexes=blastStructure(i).evalue&lt;maxE &amp; blastStructure(i).aligLen&gt;=minLen &amp; blastStructure(i).identity&gt;=minIde; <span class="comment">%Do it in this direction to lose NaNs</span>
0155     blastStructure(i).fromGenes(~indexes)=[];
0156     blastStructure(i).toGenes(~indexes)=[];
0157     blastStructure(i).evalue(~indexes)=[];
0158     blastStructure(i).identity(~indexes)=[];
0159     blastStructure(i).aligLen(~indexes)=[];
0160     blastStructure(i).bitscore(~indexes)=[];
0161     blastStructure(i).ppos(~indexes)=[];
0162 <span class="keyword">end</span>
0163 
0164 <span class="comment">%Remove all reactions from the models that have no genes encoding for them.</span>
0165 <span class="comment">%Also remove all genes that encode for no reactions. There shouldn't be any</span>
0166 <span class="comment">%but there might be mistakes</span>
0167 <span class="keyword">for</span> i=1:numel(models)
0168     [hasGenes, ~]=find(models{i}.rxnGeneMat);
0169     hasNoGenes=1:numel(models{i}.rxns);
0170     hasNoGenes(hasGenes)=[];
0171     models{i}=<a href="removeReactions.html" class="code" title="function reducedModel=removeReactions(model,rxnsToRemove,removeUnusedMets,removeUnusedGenes,removeUnusedComps)">removeReactions</a>(models{i},hasNoGenes,true,true);
0172 <span class="keyword">end</span>
0173 
0174 <span class="comment">%Create a structure that contains all genes used in the blasts in any</span>
0175 <span class="comment">%direction for each of the models in 'models' and for the new organism. The</span>
0176 <span class="comment">%first cell is for the new organism and then according to the preferred</span>
0177 <span class="comment">%order. If no such order is supplied, then according to the order in</span>
0178 <span class="comment">%'models'</span>
0179 allGenes=cell(numel(models)+1,1);
0180 <span class="keyword">if</span> isempty(preferredOrder)
0181     useOrder=modelNames;
0182 <span class="keyword">else</span>
0183     useOrder=preferredOrder;
0184 <span class="keyword">end</span>
0185 
0186 <span class="comment">%Get the corresponding indexes for those models in the 'models' structure</span>
0187 useOrderIndexes=zeros(numel(models),1);
0188 <span class="keyword">for</span> i=1:numel(models)
0189     [~, index]=ismember(models{i}.id,useOrder);
0190     useOrderIndexes(index)=i;
0191 <span class="keyword">end</span>
0192 
0193 <span class="comment">%Remove all genes from the blast structure that have no genes in the models</span>
0194 <span class="keyword">if</span> onlyGenesInModels==true
0195     modelGenes={};
0196     <span class="keyword">for</span> i=1:numel(models)
0197         modelGenes=[modelGenes;models{i}.genes];
0198     <span class="keyword">end</span>
0199     <span class="keyword">for</span> i=1:numel(blastStructure)
0200         <span class="comment">%Check to see if it should match the toId or fromId</span>
0201         <span class="keyword">if</span> strcmpi(blastStructure(i).fromId,getModelFor)
0202             I=ismember(blastStructure(i).toGenes,modelGenes);
0203         <span class="keyword">else</span>
0204             I=ismember(blastStructure(i).fromGenes,modelGenes);
0205         <span class="keyword">end</span>
0206         blastStructure(i).fromGenes(~I)=[];
0207         blastStructure(i).toGenes(~I)=[];
0208         blastStructure(i).evalue(~I)=[];
0209         blastStructure(i).identity(~I)=[];
0210         blastStructure(i).aligLen(~I)=[];
0211         blastStructure(i).bitscore(~I)=[];
0212         blastStructure(i).ppos(~I)=[];
0213         
0214         <span class="comment">%Check that no matching in blastStructure is empty. This happens if</span>
0215         <span class="comment">%no genes in the models are present in the corresponding sheet</span>
0216         <span class="keyword">if</span> isempty(blastStructure(i).fromGenes)
0217             EM=[<span class="string">'No genes in matching from '</span> blastStructure(i).fromId <span class="string">' to '</span> blastStructure(i).toId <span class="string">' are present in the corresponding model'</span>];
0218             <a href="dispEM.html" class="code" title="function dispEM(string,throwErrors,toList,trimWarnings)">dispEM</a>(EM);
0219         <span class="keyword">end</span>
0220     <span class="keyword">end</span>
0221 <span class="keyword">end</span>
0222 
0223 <span class="comment">%If only best orthologs are to be used then all other measurements are</span>
0224 <span class="comment">%deleted from the blastStructure. All code after this stays the same. This</span>
0225 <span class="comment">%means that preferred order can still matter. The best ortholog scoring is</span>
0226 <span class="comment">%based only on the E-value</span>
0227 <span class="keyword">if</span> strictness==3
0228     <span class="keyword">for</span> i=1:numel(blastStructure)
0229         keep=false(numel(blastStructure(i).toGenes),1);
0230         [allFromGenes, ~, I]=unique(blastStructure(i).fromGenes);
0231         
0232         <span class="comment">%It would be nice to get rid of this loop</span>
0233         <span class="keyword">for</span> j=1:numel(allFromGenes)
0234             allMatches=find(I==j);
0235             bestMatches=allMatches(blastStructure(i).evalue(allMatches)==min(blastStructure(i).evalue(allMatches)));
0236             
0237             <span class="comment">%Keep the best matches</span>
0238             keep(bestMatches)=true;
0239         <span class="keyword">end</span>
0240         
0241         <span class="comment">%Delete all matches that were not best matches</span>
0242         blastStructure(i).fromGenes(~keep)=[];
0243         blastStructure(i).toGenes(~keep)=[];
0244         blastStructure(i).evalue(~keep)=[];
0245         blastStructure(i).identity(~keep)=[];
0246         blastStructure(i).aligLen(~keep)=[];
0247         blastStructure(i).bitscore(~keep)=[];
0248         blastStructure(i).ppos(~keep)=[];
0249     <span class="keyword">end</span>
0250 <span class="keyword">end</span>
0251 
0252 useOrder=[{getModelFor};useOrder];
0253 
0254 <span class="keyword">for</span> i=1:numel(blastStructure)
0255     [~, toIndex]=ismember(blastStructure(i).toId,useOrder);
0256     [~, fromIndex]=ismember(blastStructure(i).fromId,useOrder);
0257     
0258     <span class="comment">%Add all genes to the corresponding list in allGenes</span>
0259     allGenes{toIndex}=[allGenes{toIndex};blastStructure(i).toGenes];
0260     allGenes{fromIndex}=[allGenes{fromIndex};blastStructure(i).fromGenes];
0261 <span class="keyword">end</span>
0262 
0263 <span class="comment">%Keep only the unique gene names</span>
0264 maxOtherGeneNr=0; <span class="comment">%Determines the dimension of the connectivity matrixes</span>
0265 <span class="keyword">for</span> i=1:numel(allGenes)
0266     allGenes{i}=unique(allGenes{i});
0267     <span class="keyword">if</span> i&gt;1
0268         <span class="keyword">if</span> numel(allGenes{i})&gt;maxOtherGeneNr
0269             maxOtherGeneNr=numel(allGenes{i});
0270         <span class="keyword">end</span>
0271     <span class="keyword">end</span>
0272 <span class="keyword">end</span>
0273 
0274 <span class="comment">%Generate a cell array of matrixes that describes how the genes in the new</span>
0275 <span class="comment">%organism map to the models. Each cell matches to the corresponding model</span>
0276 <span class="comment">%in useOrder (starting at 2 of course). First dimension is gene in new</span>
0277 <span class="comment">%organism, second which gene it is in the other organism. The second matrix</span>
0278 <span class="comment">%describes how they map back.</span>
0279 
0280 <span class="comment">%As it is now, a significant match is indicated by a 1. This could be</span>
0281 <span class="comment">%expanded to contain some kind of significance level. The first dimension</span>
0282 <span class="comment">%is still the genes in the new model.</span>
0283 
0284 allTo=cell(numel(useOrder)-1,1);
0285 allFrom=cell(numel(useOrder)-1,1);
0286 
0287 <span class="keyword">for</span> i=1:numel(useOrder)-1
0288     allTo{i}=sparse(numel(allGenes{1}),numel(allGenes{i+1}));
0289     allFrom{i}=sparse(numel(allGenes{1}),numel(allGenes{i+1}));
0290 <span class="keyword">end</span>
0291 
0292 <span class="comment">%Fill the matches to other species</span>
0293 <span class="keyword">for</span> i=1:numel(blastStructure)
0294     <span class="keyword">if</span> strcmp(blastStructure(i).toId,getModelFor)
0295         <span class="comment">%This was 'to' the new organism. They should all match so no checks</span>
0296         <span class="comment">%are being made</span>
0297         [~, a]=ismember(blastStructure(i).toGenes,allGenes{1});
0298         [~, fromModel]=ismember(blastStructure(i).fromId,useOrder);
0299         [~, b]=ismember(blastStructure(i).fromGenes,allGenes{fromModel});
0300         idx = sub2ind(size(allTo{fromModel-1}), a, b);
0301         allTo{fromModel-1}(idx)=1;
0302     <span class="keyword">else</span>
0303         <span class="comment">%This was 'from' the new organism</span>
0304         [~, a]=ismember(blastStructure(i).fromGenes,allGenes{1});
0305         [~, toModel]=ismember(blastStructure(i).toId,useOrder);
0306         [~, b]=ismember(blastStructure(i).toGenes,allGenes{toModel});
0307         idx = sub2ind(size(allFrom{toModel-1}), a, b);
0308         allFrom{toModel-1}(idx)=1;
0309     <span class="keyword">end</span>
0310 <span class="keyword">end</span>
0311 
0312 <span class="comment">%Now we have all the gene matches in a convenient way. For all the genes in</span>
0313 <span class="comment">%the new organism get the genes that should be included from other</span>
0314 <span class="comment">%organisms. If all genes should be included this simply means keep the</span>
0315 <span class="comment">%allFrom matrix as it is. If only orthologs which could be mapped in both</span>
0316 <span class="comment">%BLAST directions are to be included then only those elements are kept.</span>
0317 
0318 finalMappings=cell(numel(useOrder)-1,1);
0319 <span class="keyword">if</span> strictness==1 || strictness==3
0320     <span class="keyword">for</span> j=1:numel(allFrom)
0321         finalMappings{j}=allTo{j}~=0 &amp; allFrom{j}~=0;
0322     <span class="keyword">end</span>
0323 <span class="keyword">else</span>
0324     <span class="keyword">if</span> mapNewGenesToOld==true
0325         finalMappings=allFrom;
0326     <span class="keyword">else</span>
0327         finalMappings=allTo;
0328     <span class="keyword">end</span>
0329 <span class="keyword">end</span>
0330 
0331 <span class="comment">%Remove all genes from the mapping that are not in the models. This doesn't</span>
0332 <span class="comment">%do much if only genes in the models were used for the original mapping.</span>
0333 <span class="comment">%Also simplify the finalMapping and allGenes structures so that they only</span>
0334 <span class="comment">%contain mappings that exist</span>
0335 usedNewGenes=false(numel(allGenes{1}),1);
0336 
0337 <span class="keyword">for</span> i=1:numel(allGenes)-1
0338     <span class="comment">%First remove mappings for those genes that are not in the model</span>
0339     <span class="keyword">if</span> onlyGenesInModels==false
0340         a=ismember(allGenes{i+1},models{useOrderIndexes(i)}.genes);
0341         finalMappings{i}(:,~a)=false;
0342     <span class="keyword">end</span>
0343     
0344     <span class="comment">%Then remove unused ones and simplify</span>
0345     [a, b]=find(finalMappings{i});
0346     usedGenes=false(numel(allGenes{i+1}),1);
0347     usedNewGenes(a)=true;
0348     usedGenes(b)=true;
0349     finalMappings{i}=finalMappings{i}(:,usedGenes);
0350     allGenes{i+1}=allGenes{i+1}(usedGenes);
0351 <span class="keyword">end</span>
0352 
0353 <span class="comment">%Remove all new genes that have not been mapped to anything</span>
0354 allGenes{1}=allGenes{1}(usedNewGenes);
0355 <span class="keyword">for</span> i=1:numel(finalMappings)
0356     finalMappings{i}=finalMappings{i}(usedNewGenes,:);
0357 <span class="keyword">end</span>
0358 
0359 <span class="comment">%Now is it time to choose which reactions should be included from which</span>
0360 <span class="comment">%models. If there is a preferred order specified then each gene can only</span>
0361 <span class="comment">%result in reactions from one model, otherwise they should all be included</span>
0362 
0363 <span class="comment">%Start by simplifying the models by removing genes/reactions that are not</span>
0364 <span class="comment">%used. This is where it gets weird with complexes, especially &quot;or&quot;</span>
0365 <span class="comment">%complexes. In this step only reactions which are encoded by one single</span>
0366 <span class="comment">%gene, or where all genes should be deleted, are deleted. The info on the</span>
0367 <span class="comment">%full complex is still present in the grRules</span>
0368 
0369 <span class="keyword">for</span> i=1:numel(models)
0370     a=ismember(models{useOrderIndexes(i)}.genes,allGenes{i+1});
0371     
0372     <span class="comment">%Remove reactions that are not associated to any of the genes in</span>
0373     <span class="comment">%allGenes, thereby also keeping complexes where only for one of the</span>
0374     <span class="comment">%genes was matched</span>
0375     [rxnsToKeep,~] = find(models{useOrderIndexes(i)}.rxnGeneMat(:,a));
0376     rxnsToRemove = repmat(1,numel(models{useOrderIndexes(i)}.rxns),1);
0377     rxnsToRemove(rxnsToKeep) = 0;
0378     rxnsToRemove = find(rxnsToRemove);
0379     models{useOrderIndexes(i)}=<a href="removeReactions.html" class="code" title="function reducedModel=removeReactions(model,rxnsToRemove,removeUnusedMets,removeUnusedGenes,removeUnusedComps)">removeReactions</a>(models{useOrderIndexes(i)},rxnsToRemove,true,true,true);
0380 <span class="keyword">end</span>
0381 
0382 <span class="comment">%Since mergeModels function will be used in the end, the models are</span>
0383 <span class="comment">%simplified further by deleting genes/reactions in the order specified by</span>
0384 <span class="comment">%preferredOrder. This means that the last model will only contain reactions</span>
0385 <span class="comment">%for genes that mapped only to that model</span>
0386 
0387 allUsedGenes=false(numel(allGenes{1}),1);
0388 
0389 <span class="keyword">if</span> ~isempty(preferredOrder) &amp;&amp; numel(models)&gt;1
0390     [usedGenes, ~]=find(finalMappings{1}); <span class="comment">%All that are used in the first model in preferredOrder</span>
0391     allUsedGenes(usedGenes)=true;
0392     <span class="keyword">for</span> i=2:numel(finalMappings)
0393         [usedGenes, ~]=find(finalMappings{i});
0394         usedGenes=unique(usedGenes);
0395         a=ismember(usedGenes,find(allUsedGenes));
0396         
0397         [~, genesToDelete]=find(finalMappings{i}(usedGenes(a),:)); <span class="comment">%IMPORTANT! IS it really correct to remove all genes that map?</span>
0398         genesToDelete=unique(genesToDelete); <span class="comment">%Maybe not needed, but for clarity if nothing else</span>
0399         
0400         <span class="comment">%Remove all the genes that were already found and add the other</span>
0401         <span class="comment">%ones to allUsedGenes</span>
0402         models{useOrderIndexes(i)}=<a href="removeGenes.html" class="code" title="function reducedModel = removeGenes(model,genesToRemove,removeUnusedMets,removeBlockedRxns,standardizeRules)">removeGenes</a>(models{useOrderIndexes(i)},allGenes{i+1}(genesToDelete),true,true,false);
0403         allUsedGenes(usedGenes)=true;
0404         
0405         <span class="comment">%Remove the deleted genes from finalMappings and allGenes.</span>
0406         finalMappings{i}(:,genesToDelete)=[];
0407         allGenes{i+1}(genesToDelete)=[];
0408     <span class="keyword">end</span>
0409 <span class="keyword">end</span>
0410 
0411 <span class="comment">%Now loop through the models and update the gene associations. Genes not</span>
0412 <span class="comment">%belonging to the new organism will be renamed as 'OLD_MODELID_gene'</span>
0413 <span class="keyword">for</span> i=1:numel(models)
0414     <span class="comment">%Find all the new genes that should be used for this model</span>
0415     [newGenes, oldGenes]=find(finalMappings{i});
0416     
0417     <span class="comment">%Create a new gene list with the genes from the new organism and those</span>
0418     <span class="comment">%genes that could not be removed</span>
0419     replaceableGenes=allGenes{i+1}(unique(oldGenes));
0420     nonReplaceableGenes=setdiff(models{useOrderIndexes(i)}.genes,replaceableGenes);
0421     fullGeneList=[allGenes{1}(unique(newGenes));nonReplaceableGenes];
0422     
0423     <span class="comment">%Just to save some indexing later. This is the LAST index of</span>
0424     <span class="comment">%replaceable ones</span>
0425     nonRepStartIndex=numel(unique(newGenes));
0426     
0427     <span class="comment">%Construct a new rxnGeneMat</span>
0428     newRxnGeneMat=sparse(numel(models{useOrderIndexes(i)}.rxns),numel(fullGeneList));
0429     
0430     <span class="comment">%Now update the rxnGeneMat. This is a little tricky and could</span>
0431     <span class="comment">%probably be done in a more efficient way, but I just loop through the</span>
0432     <span class="comment">%reactions and add them one by one</span>
0433     <span class="keyword">for</span> j=1:numel(models{useOrderIndexes(i)}.rxns)
0434         <span class="comment">%Get the old genes encoding for this reaction</span>
0435         [~, oldGeneIds]=find(models{useOrderIndexes(i)}.rxnGeneMat(j,:));
0436         
0437         <span class="comment">%Update the matrix for each gene. This includes replacing one gene</span>
0438         <span class="comment">%with several new ones if there were several matches</span>
0439         <span class="keyword">for</span> k=1:numel(oldGeneIds)
0440             <span class="comment">%Match the gene to one in the gene list. This is done as a text</span>
0441             <span class="comment">%match. Could probably be done better, but I'm a little lost in</span>
0442             <span class="comment">%the indexing</span>
0443             
0444             geneName=models{useOrderIndexes(i)}.genes(oldGeneIds(k));
0445             
0446             <span class="comment">%First search in the mappable genes</span>
0447             mapIndex=find(ismember(allGenes{i+1},geneName));
0448             
0449             <span class="keyword">if</span> ~isempty(mapIndex)
0450                 <span class="comment">% add the old genes</span>
0451                 hitGenes.oldGenes = [hitGenes.oldGenes, {geneName}];
0452                 
0453                 <span class="comment">%Get the new genes for that gene</span>
0454                 a=find(finalMappings{i}(:,mapIndex));
0455                 
0456                 <span class="comment">%Find the positions of these genes in the final gene list</span>
0457                 [~, b]=ismember(allGenes{1}(a),fullGeneList);
0458                 
0459                 <span class="comment">%Update the matrix</span>
0460                 newRxnGeneMat(j,b)=1;
0461                 
0462                 <span class="comment">%Update the grRules string. This is tricky, but I hope that</span>
0463                 <span class="comment">%it's ok to replace the old gene name with the new one and</span>
0464                 <span class="comment">%add ') or (' if there were several matches. Be sure of</span>
0465                 <span class="comment">%this!</span>
0466                 repString=fullGeneList{b(1)};
0467                 <span class="keyword">if</span> numel(b)&gt;1
0468                     <span class="keyword">for</span> l=2:numel(b)
0469                         repString=[repString <span class="string">' or '</span> fullGeneList{b(l)}];
0470                     <span class="keyword">end</span>
0471                     repString=[<span class="string">'('</span> repString <span class="string">')'</span>];
0472                 <span class="keyword">end</span>
0473                 
0474                 <span class="comment">% add the new matched genes</span>
0475                 hitGenes.newGenes = [hitGenes.newGenes, {repString}];
0476                 
0477                 <span class="comment">%Use regexprep instead of strrep to prevent partial matches</span>
0478                 models{useOrderIndexes(i)}.grRules{j}=regexprep(models{useOrderIndexes(i)}.grRules{j},[<span class="string">'(^|\s|\()'</span> geneName{1} <span class="string">'($|\s|\))'</span>],[<span class="string">'$1'</span> repString <span class="string">'$2'</span>]);
0479             <span class="keyword">else</span>
0480                 <span class="comment">%Then search in the non-replaceable genes. There could only</span>
0481                 <span class="comment">%be one match here</span>
0482                 index=find(ismember(nonReplaceableGenes,geneName));
0483                 
0484                 <span class="comment">%Update the matrix</span>
0485                 newRxnGeneMat(j,nonRepStartIndex+index)=1;
0486                 
0487                 models{useOrderIndexes(i)}.grRules{j}=strrep(models{useOrderIndexes(i)}.grRules{j},geneName{1},strcat(<span class="string">'OLD_'</span>,models{useOrderIndexes(i)}.id,<span class="string">'_'</span>,geneName{1}));
0488             <span class="keyword">end</span>
0489         <span class="keyword">end</span>
0490     <span class="keyword">end</span>
0491     
0492     <span class="comment">%Add the new list of genes</span>
0493     models{useOrderIndexes(i)}.rxnGeneMat=newRxnGeneMat;
0494     <span class="keyword">if</span> ~isempty(nonReplaceableGenes)
0495         models{useOrderIndexes(i)}.genes=[allGenes{1}(unique(newGenes));strcat(<span class="string">'OLD_'</span>,models{useOrderIndexes(i)}.id,<span class="string">'_'</span>,nonReplaceableGenes)];
0496     <span class="keyword">else</span>
0497         models{useOrderIndexes(i)}.genes=allGenes{1}(unique(newGenes));
0498     <span class="keyword">end</span>
0499     <span class="keyword">if</span> isfield(models{useOrderIndexes(i)},<span class="string">'geneComps'</span>)
0500         geneComps=models{useOrderIndexes(i)}.geneComps(1);
0501         models{useOrderIndexes(i)}.geneComps=zeros(numel(models{useOrderIndexes(i)}.genes),1);
0502         <span class="comment">%Assume that all genes are in the same compartment, and this</span>
0503         <span class="comment">%compartment is specified for the first gene</span>
0504         models{useOrderIndexes(i)}.geneComps(:)=geneComps;
0505     <span class="keyword">end</span>
0506 <span class="keyword">end</span>
0507 
0508 <span class="comment">%Now merge the models. All information should be correct except for 'or'</span>
0509 <span class="comment">%complexes</span>
0510 draftModel=<a href="mergeModels.html" class="code" title="function model=mergeModels(models,metParam,supressWarnings)">mergeModels</a>(models,<span class="string">'metNames'</span>);
0511 
0512 <span class="comment">%Change name of the resulting model</span>
0513 draftModel.id=getModelFor;
0514 name=<span class="string">'Generated by getModelFromHomology using '</span>;
0515 <span class="keyword">for</span> i=1:numel(models)
0516     <span class="keyword">if</span> i&lt;numel(models)
0517         name=[name models{i}.id <span class="string">', '</span>];
0518     <span class="keyword">else</span>
0519         name=[name models{i}.id];
0520     <span class="keyword">end</span>
0521 <span class="keyword">end</span>
0522 draftModel.name=name;
0523 draftModel.rxnNotes=cell(length(draftModel.rxns),1);
0524 draftModel.rxnNotes(:)={<span class="string">'Included by getModelFromHomology'</span>};
0525 draftModel.rxnConfidenceScores=NaN(length(draftModel.rxns),1);
0526 draftModel.rxnConfidenceScores(:)=2;
0527 draftModel=<a href="deleteUnusedGenes.html" class="code" title="function reducedModel=deleteUnusedGenes(model,verbose)">deleteUnusedGenes</a>(draftModel,0);
0528 <span class="comment">%Standardize grRules and notify if problematic grRules are found</span>
0529 [draftModel.grRules,draftModel.rxnGeneMat]=<a href="standardizeGrRules.html" class="code" title="function [grRules,rxnGeneMat,indexes2check] = standardizeGrRules(model,embedded)">standardizeGrRules</a>(draftModel,false);
0530 <span class="keyword">end</span></pre></div>
<hr><address>Generated by <strong><a href="http://www.artefact.tk/software/matlab/m2html/" title="Matlab Documentation in HTML">m2html</a></strong> &copy; 2005</address>
</body>
</html>